// @dbcube/cli/src/lib/DBCubeLogger.js
class DBCubeLogger {
    constructor(title) {
        this.logs = [];
        this.maxLogs = 3;
        this.title = title;
        this.isWaiting = false;
        this.animationInterval = null;
        this.currentFace = 0;
        this.faces = ['● ‿ ●', '● ‿ ●', '● ‿ <', '● ‿ <', '● ‿ ●', '● ‿ ●', '> ‿ ●', '> ‿ ●', '● ‿ ●', '● ‿ ●', '> ‿ <', '> ‿ <', '● ‿ ●', '● ‿ ●', '> ‿ <', '> ‿ <'];
        
        // Registrar globalmente para que LoggerConsole pueda acceder
        global.dbcubeLogger = this;
    }

    // Función para limpiar solo el área superior de la consola
    clearConsole() {
        // Solo limpiar si estamos en modo división de consola
        if (global.loggerConsoleInstances && global.loggerConsoleInstances.top && global.loggerConsoleInstances.top.isActive) {
            // Usar el método clear interceptado que solo afecta al área superior
            console.clear();
        } else {
            // Comportamiento normal si no hay división
            console.clear();
        }
    }

    // Función para dibujar el cubo con logs
    drawCube() {
        this.clearConsole();
        console.log(this.title);
        
        // Mostrar logs anteriores (máximo 3)
        for (let i = 0; i < this.logs.length - 3; i++) {
            const log = this.logs[i];
            if (i == 0) console.log('╭───────╮');
            console.log('│───────│  ' + log.message);
        }
        
        if (this.logs.length == 1) {
            console.log('╭───────╮');
            console.log('│───────│');
            console.log('│───────│');
        }
        
        if (this.logs.length == 2) {
            const penPastLog = this.logs[this.logs.length - 2];
            console.log('╭───────╮');
            console.log('│───────│');
            console.log('│───────│  ' + penPastLog.message);
        }
        
        if (this.logs.length == 3) {
            const penPastLog = this.logs[this.logs.length - 3];
            const penPastLog2 = this.logs[this.logs.length - 2];
            console.log('╭───────╮');
            console.log('│───────│  ' + penPastLog.message);
            console.log('│───────│  ' + penPastLog2.message);
        }
        
        // Mostrar el cubo con cara en la última posición
        if (this.logs.length > 0) {
            const lastLog = this.logs[this.logs.length - 1];
            const face = this.isWaiting ? this.faces[this.currentFace] : lastLog.face;
            console.log('│ ' + face + ' │  ' + lastLog.message);
            console.log('╰───────╯');
        }
    }

    // Función para agregar un log
    addLog(message, face) {
        const log = { message, face };
        this.logs.push(log);
        
        // Mantener máximo 3 logs
        if (this.logs.length > this.maxLogs) {
            this.logs.shift();
        }
        
        if (!this.isWaiting) {
            this.drawCube();
        }
    }

    // Función para logs normales
    info(message) {
        this.addLog(message, '● ‿ ●');
    }

    // Función para logs de error
    error(message) {
        this.addLog(message, 'x _ x');
    }

    // Función para comenzar la animación de espera
    wait() {
        this.isWaiting = true;
        this.currentFace = 0;
        
        this.animationInterval = setInterval(() => {
            this.currentFace = (this.currentFace + 1) % this.faces.length;
            this.drawCube();
        }, 500); // Cambia cada 500ms
    }

    // Función para detener la animación
    continue() {
        this.isWaiting = false;
        if (this.animationInterval) {
            clearInterval(this.animationInterval);
            this.animationInterval = null;
        }
        this.drawCube();
    }
}

module.exports = DBCubeLogger;
