// @dbcube/cli/src/lib/LoggerConsole.js
class LoggerConsole {
    constructor(topLines = null) {
        this.topLines = topLines;
        this.isTopConsole = topLines !== null && topLines !== undefined;
        this.isActive = false;
        this.originalConsole = {};
        this.buffer = [];
        this.bottomBuffer = [];
        this.topBuffer = [];
        this.terminalHeight = process.stdout.rows || 24;
        this.terminalWidth = process.stdout.columns || 80;
        
        // Compartir instancias globalmente
        if (!global.loggerConsoleInstances) {
            global.loggerConsoleInstances = {
                top: null,
                bottom: null,
                sharedState: {
                    topBuffer: [],
                    bottomBuffer: [],
                    activeInstance: null
                }
            };
        }
        
        // Registrar la instancia
        if (this.isTopConsole) {
            global.loggerConsoleInstances.top = this;
        } else {
            global.loggerConsoleInstances.bottom = this;
        }
    }

    start() {
        if (this.isActive) return;
        
        this.isActive = true;
        global.loggerConsoleInstances.sharedState.activeInstance = this;
        
        // Interceptar console methods solo para la instancia activa
        if (this.isTopConsole) {
            this.interceptConsole();
        }
        
        // Configurar el terminal
        this.setupTerminal();
        
        // Render inicial
        this.render();
    }

    end() {
        if (!this.isActive) return;
        
        this.isActive = false;

        // Restaurar console original
        if (this.isTopConsole) {
            this.restoreConsole();
        }
        
        // Limpiar terminal
        this.cleanup();
        
        
        global.loggerConsoleInstances.sharedState.activeInstance = null;
    }

    setupTerminal() {
        // Ocultar cursor
        process.stdout.write('\x1B[?25l');
        
        // Limpiar pantalla
        process.stdout.write('\x1B[2J\x1B[H');
        
        // Configurar el área de trabajo
        this.reserveArea();
    }

    reserveArea() {
        if (this.isTopConsole) {
            // Reservar líneas superiores
            for (let i = 0; i < this.topLines; i++) {
                process.stdout.write('\n');
            }
            // Posicionar cursor en la primera línea reservada
            process.stdout.write(`\x1B[${this.topLines}A`);
        } else {
            // Posicionar cursor en la parte inferior
            const bottomStartLine = (global.loggerConsoleInstances.top?.topLines || 0) + 2;
            process.stdout.write(`\x1B[${bottomStartLine};1H`);
        }
    }

    interceptConsole() {
        // Guardar métodos originales
        this.originalConsole.log = console.log;
        this.originalConsole.info = console.info;
        this.originalConsole.error = console.error;
        this.originalConsole.warn = console.warn;
        this.originalConsole.clear = console.clear;

        // Interceptar console.log para la consola superior
        console.log = (...args) => {
            if (this.isActive && this.isTopConsole) {
                this.addToTopBuffer(args.join(' '));
                this.render();
            }
        };

        console.info = (...args) => {
            if (this.isActive && this.isTopConsole) {
                this.addToTopBuffer(args.join(' '));
                this.render();
            }
        };

        console.error = (...args) => {
            if (this.isActive && this.isTopConsole) {
                this.addToTopBuffer(args.join(' '));
                this.render();
            }
        };

        console.warn = (...args) => {
            if (this.isActive && this.isTopConsole) {
                this.addToTopBuffer(args.join(' '));
                this.render();
            }
        };

        console.clear = () => {
            if (this.isActive && this.isTopConsole) {
                this.clearTopBuffer();
                this.render();
            }
        };
    }

    restoreConsole() {
        console.log = this.originalConsole.log;
        console.info = this.originalConsole.info;
        console.error = this.originalConsole.error;
        console.warn = this.originalConsole.warn;
        console.clear = this.originalConsole.clear;
    }

    addToTopBuffer(message) {
        global.loggerConsoleInstances.sharedState.topBuffer.push(message);
        
        // Mantener solo las líneas necesarias
        if (global.loggerConsoleInstances.sharedState.topBuffer.length > this.topLines) {
            global.loggerConsoleInstances.sharedState.topBuffer.shift();
        }
    }

    addToBottomBuffer(message) {
        global.loggerConsoleInstances.sharedState.bottomBuffer.push(message);
    }

    clearTopBuffer() {
        global.loggerConsoleInstances.sharedState.topBuffer = [];
    }

    render() {
        if (!this.isActive) return;
        
        if (this.isTopConsole) {
            this.renderTopArea();
        } else {
            this.renderBottomArea();
        }
    }

    renderTopArea() {
        // Guardar posición del cursor
        process.stdout.write('\x1B[s');
        
        // Ir al inicio del área superior
        process.stdout.write('\x1B[1;1H');
        
        // Limpiar área superior
        for (let i = 0; i < this.topLines; i++) {
            process.stdout.write('\x1B[2K'); // Limpiar línea
            if (i < this.topLines - 1) process.stdout.write('\n');
        }
        
        // Volver al inicio del área superior
        process.stdout.write('\x1B[1;1H');
        
        // Renderizar buffer superior
        const topBuffer = global.loggerConsoleInstances.sharedState.topBuffer;
        for (let i = 0; i < Math.min(topBuffer.length, this.topLines); i++) {
            process.stdout.write(topBuffer[i]);
            if (i < Math.min(topBuffer.length, this.topLines) - 1) {
                process.stdout.write('\n');
            }
        }
        
        // Restaurar posición del cursor
        process.stdout.write('\x1B[u');
    }

    renderBottomArea() {
        const topLines = global.loggerConsoleInstances.top?.topLines || 0;
        const separatorLine = topLines + 1;
        const bottomStartLine = topLines + 2;
        
        // Dibujar separador
        process.stdout.write(`\x1B[${separatorLine};1H`);
        process.stdout.write('\x1B[2K');
        process.stdout.write('─'.repeat(this.terminalWidth));
        
        // Posicionar cursor en área inferior
        process.stdout.write(`\x1B[${bottomStartLine};1H`);
    }

    cleanup() {
        // Mostrar cursor
        process.stdout.write('\x1B[?25h');
        
        // Ir al final de la pantalla
        process.stdout.write(`\x1B[${this.terminalHeight};1H`);
        
        // Limpiar pantalla si es necesario
        if (this.isTopConsole) {
            process.stdout.write('\x1B[2J\x1B[H');
        }
    }

    // Método para enviar logs desde la consola inferior a la superior
    sendToTop(message, type = 'info') {
        if (global.loggerConsoleInstances.top && global.loggerConsoleInstances.top.isActive) {
            // Simular llamada al método del DBCubeLogger
            if (global.dbcubeLogger) {
                switch (type) {
                    case 'info':
                        global.dbcubeLogger.info(message);
                        break;
                    case 'error':
                        global.dbcubeLogger.error(message);
                        break;
                    case 'wait':
                        global.dbcubeLogger.wait();
                        break;
                    case 'continue':
                        global.dbcubeLogger.continue();
                        break;
                }
            }
        }
    }

    // Método estático para facilitar el envío de logs
    static sendToTop(message, type = 'info') {
        if (global.loggerConsoleInstances.bottom) {
            global.loggerConsoleInstances.bottom.sendToTop(message, type);
        }
    }
}

module.exports = LoggerConsole;
