// @dbcube/cli/src/lib/ProgressIndicator.js
const chalk = require('chalk');
const ora = require('ora');
const fs = require('fs');
const path = require('path');

class ProgressIndicator {
    constructor() {
        this.activeSpinner = null;
        this.startTime = null;
    }

    /**
     * Inicia el indicador de progreso principal
     * @param {string} message - Mensaje principal
     */
    start(message) {
        console.log(`\n${chalk.cyan('●')} ${chalk.bold(message)}`);
        this.startTime = Date.now();
    }

    /**
     * Muestra progreso para una operación específica con tabla
     * @param {string} tableName - Nombre de la tabla
     * @param {string} operation - Tipo de operación (refresh, seeder, trigger)
     * @param {string} databaseName - Nombre de la base de datos
     */
    async showTableProgress(tableName, operation, databaseName = '') {
        const startTime = Date.now();
        const maxDots = 50;
        const operationText = this.getOperationText(operation);

        // Formatear el nombre para que tenga un ancho fijo
        const displayName = databaseName ? `${databaseName}.${tableName}` : tableName;
        const truncatedName = displayName.length > 30 ?
            displayName.substring(0, 27) + '...' :
            displayName.padEnd(30);

        // Simular progreso con puntos
        return new Promise((resolve) => {
            process.stdout.write(`${chalk.gray('├─')} ${truncatedName} `);

            let dotCount = 0;
            const interval = setInterval(() => {
                if (dotCount < maxDots) {
                    process.stdout.write(chalk.gray('.'));
                    dotCount++;
                } else {
                    clearInterval(interval);
                    const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
                    process.stdout.write(` ${chalk.green('✓')} ${chalk.gray(elapsed + 's')}\n`);
                    resolve();
                }
            }, 20); // Velocidad de los puntos
        });
    }

    /**
     * Muestra progreso para múltiples tablas
     * @param {Array} tables - Array de nombres de tablas
     * @param {string} operation - Tipo de operación
     * @param {string} databaseName - Nombre de la base de datos
     * @param {Function} executor - Función que ejecuta la operación para cada tabla
     */
    async showMultipleTablesProgress(tables, operation, databaseName, executor) {
        const operationText = this.getOperationText(operation);
        console.log(`  ${chalk.blue('┌─')} ${operationText} en ${chalk.bold(databaseName)}:`);

        for (let i = 0; i < tables.length; i++) {
            const table = tables[i];
            const isLast = i === tables.length - 1;

            await this.showTableProgress(table, operation, '');

            if (executor) {
                await executor(table);
            }
        }

        console.log(`${chalk.blue('└─')} ${chalk.green('Completado')}`);
    }

    /**
     * Muestra progreso simple para operaciones generales
     * @param {string} message - Mensaje a mostrar
     * @param {Function} executor - Función a ejecutar
     */
    async showProgress(message, executor) {
        const spinner = ora(message).start();
        const startTime = Date.now();

        try {
            const result = await executor();
            const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
            spinner.succeed(`${message} ${chalk.gray(`(${elapsed}s)`)}`);
            return result;
        } catch (error) {
            spinner.fail(`${message} ${chalk.red('falló')}`);
            throw error;
        }
    }

    /**
     * Muestra el resumen final
     * @param {string} operation - Tipo de operación
     * @param {number} count - Cantidad de elementos procesados
     * @param {number} databases - Cantidad de bases de datos
     */
    showSummary(operation, count, databases) {
        const totalTime = this.startTime ? ((Date.now() - this.startTime) / 1000).toFixed(1) : '0.0';
        const operationText = this.getOperationText(operation);

        console.log(`\n${chalk.green('🎉')} ${chalk.bold('Resumen:')}`);
        console.log(`  ${chalk.gray('├─')} ${operationText}: ${chalk.cyan(count)} elementos`);
        console.log(`  ${chalk.gray('├─')} Bases de datos: ${chalk.cyan(databases)}`);
        console.log(`  ${chalk.gray('└─')} Tiempo total: ${chalk.cyan(totalTime + 's')}`);
    }

    /**
     * Inicia el cronómetro
     */
    startTimer() {
        this.startTime = Date.now();
    }

    /**
     * Muestra error con formato
     * @param {string} message - Mensaje de error
     * @param {Error} error - Error objeto
     */
    showError(message, error) {
        console.log(`\n${chalk.red('❌')} ${chalk.bold(message)}`);
        if (error && error.message) {
            console.log(`  ${chalk.gray('└─')} ${chalk.red(error.message)}`);
        }
    }

    /**
     * Obtiene el texto de la operación
     * @param {string} operation - Tipo de operación
     * @returns {string}
     */
    getOperationText(operation) {
        const operations = {
            'refresh': 'Refrescando tablas',
            'seeder': 'Ejecutando seeders',
            'trigger': 'Ejecutando triggers'
        };
        return operations[operation] || operation;
    }

    /**
     * Extrae nombres de tablas reales de archivos .cube
     * @param {string[]} cubeFiles - Array de rutas de archivos .cube
     * @returns {string[]} Array de nombres de tablas
     */
    extractTableNamesFromCubes(cubeFiles) {
        const tableNames = [];

        for (const filePath of cubeFiles) {
            try {
                const content = fs.readFileSync(filePath, 'utf8');

                // Patrón principal: @meta({ name: "nombre_tabla"; }) o @meta({ name: 'nombre_tabla'; })
                const metaMatch = content.match(/@meta\s*\(\s*\{\s*name\s*:\s*["']([^"']+)["']\s*;?\s*[^}]*\}\s*\)/);
                if (metaMatch) {
                    tableNames.push(metaMatch[1]);
                    continue;
                }

                // Patrón alternativo: @table("nombre_tabla") o @table('nombre_tabla')
                const tableMatch = content.match(/@table\(\s*["']([^"']+)["']\s*\)/);
                if (tableMatch) {
                    tableNames.push(tableMatch[1]);
                    continue;
                }

                // Patrón 3: table: "nombre_tabla" o table: 'nombre_tabla'
                const tableMatch2 = content.match(/table\s*:\s*["']([^"']+)["']/);
                if (tableMatch2) {
                    tableNames.push(tableMatch2[1]);
                    continue;
                }

                // Fallback: usar el nombre del archivo sin la extensión .cube
                const fileName = path.basename(filePath, '.table.cube');
                if (fileName && fileName !== 'table') {
                    // Si el archivo tiene un número al principio, quitarlo
                    const cleanName = fileName.replace(/^\d+[-_]?/, '');
                    if (cleanName) {
                        tableNames.push(cleanName);
                    }
                }

            } catch (error) {
                // Si no se puede leer el archivo, usar el nombre del archivo
                const fileName = path.basename(filePath, '.table.cube');
                const cleanName = fileName.replace(/^\d+[-_]?/, '');
                if (cleanName && cleanName !== 'table') {
                    tableNames.push(cleanName);
                }
            }
        }

        return [...new Set(tableNames)]; // Eliminar duplicados
    }

    /**
     * Extrae nombres de seeders de archivos .cube
     * @param {string[]} cubeFiles - Array de rutas de archivos .cube
     * @returns {string[]} Array de nombres de seeders
     */
    extractSeederNamesFromCubes(cubeFiles) {
        const seederNames = [];

        for (const filePath of cubeFiles) {
            try {
                const content = fs.readFileSync(filePath, 'utf8');

                // Buscar secciones de seeders en el archivo
                const seederMatches = content.match(/@seeder\s*\(\s*["']([^"']+)["']\s*\)/g);
                if (seederMatches) {
                    for (const match of seederMatches) {
                        const seederName = match.match(/@seeder\s*\(\s*["']([^"']+)["']\s*\)/)[1];
                        seederNames.push(seederName);
                    }
                } else {
                    // Si no hay seeders específicos, crear uno basado en el nombre de la tabla
                    const tableNames = this.extractTableNamesFromCubes([filePath]);
                    if (tableNames.length > 0) {
                        seederNames.push(`${tableNames[0]}Seeder`);
                    }
                }

            } catch (error) {
                // Fallback: usar nombre del archivo
                const fileName = path.basename(filePath, '.table.cube');
                const cleanName = fileName.replace(/^\d+[-_]?/, '');
                if (cleanName && cleanName !== 'table') {
                    seederNames.push(`${cleanName}Seeder`);
                }
            }
        }

        return seederNames;
    }

    /**
     * Extrae nombres de triggers de archivos .cube
     * @param {string[]} cubeFiles - Array de rutas de archivos .cube
     * @returns {string[]} Array de nombres de triggers
     */
    extractTriggerNamesFromCubes(cubeFiles) {
        const triggerNames = [];

        for (const filePath of cubeFiles) {
            try {
                const content = fs.readFileSync(filePath, 'utf8');

                // Buscar triggers en el archivo
                const triggerMatches = content.match(/@trigger\s*\(\s*["']([^"']+)["']\s*\)/g);
                if (triggerMatches) {
                    for (const match of triggerMatches) {
                        const triggerName = match.match(/@trigger\s*\(\s*["']([^"']+)["']\s*\)/)[1];
                        triggerNames.push(triggerName);
                    }
                } else {
                    // Si no hay triggers específicos, crear algunos basados en la tabla
                    const tableNames = this.extractTableNamesFromCubes([filePath]);
                    if (tableNames.length > 0) {
                        const tableName = tableNames[0];
                        triggerNames.push(`before_insert_${tableName}`);
                    }
                }

            } catch (error) {
                // Fallback: usar nombre del archivo
                const fileName = path.basename(filePath, '.table.cube');
                const cleanName = fileName.replace(/^\d+[-_]?/, '');
                if (cleanName && cleanName !== 'table') {
                    triggerNames.push(`before_insert_${cleanName}`);
                }
            }
        }

        return triggerNames;
    }

    /**
     * Limpia la pantalla y muestra encabezado
     * @param {string} title - Título principal
     * @param {string} emoji - Emoji a mostrar
     */
    showHeader(title, emoji = '🔄') {
        console.clear();
        console.log(`\n${emoji} ${chalk.green.bold(title)}`);
        console.log(chalk.gray('─'.repeat(50)));
        this.startTime = Date.now(); // Iniciar cronómetro aquí
    }
}

module.exports = ProgressIndicator;
