// @dbcube/core/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Arquitecture: () => Arquitecture,
  Binary: () => Binary,
  ComputedFieldProcessor: () => ComputedFieldProcessor,
  Config: () => Config,
  DaemonClient: () => DaemonClient,
  DbConfig: () => DbConfig,
  Engine: () => Engine,
  FileLogger: () => FileLogger,
  QueryEngine: () => QueryEngine,
  TableProcessor: () => TableProcessor,
  TriggerProcessor: () => TriggerProcessor
});
module.exports = __toCommonJS(index_exports);

// src/lib/Engine.ts
var import_path4 = __toESM(require("path"));

// src/lib/Arquitecture.ts
var os = __toESM(require("os"));
var Arquitecture = class {
  systemInfo;
  constructor() {
    this.systemInfo = this.detectSystemInfo();
  }
  /**
   * Detecta información completa del sistema
   */
  detectSystemInfo() {
    return {
      platform: os.platform(),
      arch: os.arch(),
      release: os.release(),
      type: os.type(),
      endianness: os.endianness(),
      cpus: os.cpus().length
    };
  }
  /**
   * Obtiene la plataforma normalizada
   */
  getPlatform() {
    const platform2 = this.systemInfo.platform;
    switch (platform2) {
      case "win32":
        return "windows";
      case "darwin":
        return "macos";
      case "linux":
        return "linux";
      case "freebsd":
        return "freebsd";
      case "openbsd":
        return "openbsd";
      case "sunos":
        return "solaris";
      default:
        return platform2;
    }
  }
  /**
   * Obtiene la arquitectura normalizada
   */
  getArchitecture() {
    const arch2 = this.systemInfo.arch;
    switch (arch2) {
      case "x64":
        return "x86_64";
      case "x32":
      case "ia32":
        return "i686";
      case "arm64":
        return "aarch64";
      case "arm":
        return "armv7";
      case "ppc64":
        return "powerpc64";
      case "ppc":
        return "powerpc";
      case "s390x":
        return "s390x";
      case "mips":
        return "mips";
      case "mipsel":
        return "mipsel";
      default:
        return arch2;
    }
  }
  /**
   * Genera el nombre del binario basado en la arquitectura
   */
  getBinaryName(baseName) {
    const platform2 = this.getPlatform();
    const arch2 = this.getArchitecture();
    const extension = platform2 === "windows" ? ".exe" : "";
    return `${baseName}-${platform2}-${arch2}${extension}`;
  }
  /**
   * Obtiene información completa del sistema
   */
  getSystemInfo() {
    return { ...this.systemInfo };
  }
  /**
   * Obtiene el triple de destino (target triple) para Rust
   */
  getRustTargetTriple() {
    const platform2 = this.getPlatform();
    const arch2 = this.getArchitecture();
    const targetMap = {
      "linux-x86_64": "x86_64-unknown-linux-gnu",
      "linux-i686": "i686-unknown-linux-gnu",
      "linux-aarch64": "aarch64-unknown-linux-gnu",
      "linux-armv7": "armv7-unknown-linux-gnueabihf",
      "macos-x86_64": "x86_64-apple-darwin",
      "macos-aarch64": "aarch64-apple-darwin",
      "windows-x86_64": "x86_64-pc-windows-msvc",
      "windows-i686": "i686-pc-windows-msvc",
      "windows-aarch64": "aarch64-pc-windows-msvc",
      "freebsd-x86_64": "x86_64-unknown-freebsd"
    };
    const key = `${platform2}-${arch2}`;
    return targetMap[key] || `${arch2}-unknown-${platform2}`;
  }
  /**
   * Muestra información detallada del sistema
   */
  printSystemInfo() {
    console.log("\u{1F5A5}\uFE0F  System Information:");
    console.log("\u251C\u2500 Platform:", this.getPlatform());
    console.log("\u251C\u2500 Arquitecture:", this.getArchitecture());
    console.log("\u251C\u2500 OS Type:", this.systemInfo.type);
    console.log("\u251C\u2500 OS Release:", this.systemInfo.release);
    console.log("\u251C\u2500 Endianness:", this.systemInfo.endianness);
    console.log("\u251C\u2500 CPUs:", this.systemInfo.cpus);
    console.log("\u2514\u2500 Rust Target:", this.getRustTargetTriple());
  }
};

// src/lib/Donwloader.ts
var fs = __toESM(require("fs"));
var path = __toESM(require("path"));
var os2 = __toESM(require("os"));
var import_follow_redirects = __toESM(require("follow-redirects"));
var unzipper = __toESM(require("unzipper"));
var import_ora = __toESM(require("ora"));
var import_chalk = __toESM(require("chalk"));
var import_url = require("url");
var import_path = require("path");
var import_meta = {};
var { https } = import_follow_redirects.default;
var Downloader = class {
  static mainSpinner = null;
  static currentSpinner = null;
  static VERSION_URLS = {
    query: "https://raw.githubusercontent.com/Dbcube/binaries/main/query-engines.json",
    schema: "https://raw.githubusercontent.com/Dbcube/binaries/main/schema-engines.json",
    sqlite: "https://raw.githubusercontent.com/Dbcube/binaries/main/sqlite-engines.json"
  };
  /**
   * Fetch latest version from GitHub
   */
  static async fetchLatestVersion(prefix) {
    const url = this.VERSION_URLS[prefix];
    return new Promise((resolve6, reject) => {
      https.get(url, (response) => {
        let data = "";
        response.on("data", (chunk) => {
          data += chunk;
        });
        response.on("end", () => {
          try {
            const versions = JSON.parse(data);
            if (versions && versions.length > 0) {
              resolve6(versions[0].version);
            } else {
              reject(new Error("No versions found"));
            }
          } catch (error) {
            reject(error);
          }
        });
        response.on("error", reject);
      }).on("error", reject);
    });
  }
  /**
   * Extract version from binary filename
   * Example: schema-engine-v1.0.0-windows-x64.exe -> v1.0.0
   */
  static extractVersionFromFilename(filename) {
    const match = filename.match(/-(v\d+\.\d+\.\d+)-/);
    return match ? match[1] : null;
  }
  /**
   * Get local version of installed binary
   */
  static getLocalVersion(binDir, prefix) {
    try {
      const files = fs.readdirSync(binDir);
      const binaryPattern = new RegExp(`^${prefix}-engine-v`);
      const binaryFile = files.find((f) => binaryPattern.test(f));
      if (binaryFile) {
        return this.extractVersionFromFilename(binaryFile);
      }
    } catch (error) {
    }
    return null;
  }
  /**
   * Compare versions (returns true if remote is newer)
   */
  static isNewerVersion(localVersion, remoteVersion) {
    if (!localVersion) return true;
    const cleanLocal = localVersion.replace(/^v/, "");
    const cleanRemote = remoteVersion.replace(/^v/, "");
    const localParts = cleanLocal.split(".").map(Number);
    const remoteParts = cleanRemote.split(".").map(Number);
    for (let i = 0; i < 3; i++) {
      if (remoteParts[i] > localParts[i]) return true;
      if (remoteParts[i] < localParts[i]) return false;
    }
    return false;
  }
  /**
   * Delete old binary files
   */
  static deleteOldBinary(binDir, prefix) {
    try {
      const files = fs.readdirSync(binDir);
      const binaryPattern = new RegExp(`^${prefix}-engine-`);
      files.forEach((file) => {
        if (binaryPattern.test(file)) {
          const filePath = path.join(binDir, file);
          try {
            fs.unlinkSync(filePath);
            console.log(`\u{1F5D1}\uFE0F  Deleted old binary: ${file}`);
          } catch (err) {
            console.warn(`\u26A0\uFE0F  Could not delete: ${file}`);
          }
        }
      });
    } catch (error) {
    }
  }
  static get(prefix) {
    const arch2 = new Arquitecture();
    const platform2 = arch2.getPlatform();
    const architecture = arch2.getArchitecture();
    const platformMap = {
      windows: "windows",
      linux: "linux",
      darwin: "macos"
    };
    const archMap = {
      x86_64: "x64",
      aarch64: "arm64"
    };
    const plat = platformMap[platform2];
    const archSuffix = archMap[architecture];
    if (plat && archSuffix) {
      const baseName = `${prefix}-engine-${plat}-${archSuffix}`;
      const binaryName = platform2 === "windows" ? `${baseName}.exe` : baseName;
      const url = `https://github.com/Dbcube/binaries/releases/download/${prefix}-engine/${prefix}-engine-latest-${plat}-${archSuffix}.zip`;
      return {
        name: binaryName,
        url,
        version: "latest",
        query_engine: binaryName,
        schema_engine: `${prefix}-engine-${plat}-${archSuffix}${platform2 === "windows" ? ".exe" : ""}`
      };
    }
    return {
      name: "",
      url: "",
      version: "",
      query_engine: "",
      schema_engine: ""
    };
  }
  /**
   * Get binary info with actual version from GitHub
   */
  static async getWithVersion(prefix) {
    const arch2 = new Arquitecture();
    const platform2 = arch2.getPlatform();
    const architecture = arch2.getArchitecture();
    const platformMap = {
      windows: "windows",
      linux: "linux",
      darwin: "macos"
    };
    const archMap = {
      x86_64: "x64",
      aarch64: "arm64"
    };
    const plat = platformMap[platform2];
    const archSuffix = archMap[architecture];
    if (plat && archSuffix) {
      const version = await this.fetchLatestVersion(prefix);
      const baseName = `${prefix}-engine-${version}-${plat}-${archSuffix}`;
      const binaryName = platform2 === "windows" ? `${baseName}.exe` : baseName;
      const url = `https://github.com/Dbcube/binaries/releases/download/${prefix}-engine/${prefix}-engine-${version}-${plat}-${archSuffix}.zip`;
      return {
        name: binaryName,
        url,
        version,
        query_engine: binaryName,
        schema_engine: binaryName
      };
    }
    return {
      name: "",
      url: "",
      version: "",
      query_engine: "",
      schema_engine: ""
    };
  }
  static async download(targetDir) {
    const binDir = targetDir || this.getDefaultBinDir();
    fs.mkdirSync(binDir, { recursive: true });
    this.mainSpinner = (0, import_ora.default)({
      text: import_chalk.default.blue("Verificando versiones de binarios..."),
      spinner: "dots12"
    }).start();
    const binariesToProcess = [];
    for (const prefix of ["query", "schema", "sqlite"]) {
      try {
        const localVersion = this.getLocalVersion(binDir, prefix);
        const remoteVersion = await this.fetchLatestVersion(prefix);
        const needsUpdate = this.isNewerVersion(localVersion, remoteVersion);
        binariesToProcess.push({
          prefix,
          needsUpdate,
          localVersion,
          remoteVersion
        });
        if (needsUpdate) {
          console.log(`
\u{1F4E6} ${prefix}-engine: ${localVersion || "not installed"} \u2192 ${remoteVersion}`);
        } else if (localVersion) {
          console.log(`
\u2705 ${prefix}-engine: ${localVersion} (up to date)`);
        }
      } catch (error) {
        console.warn(`\u26A0\uFE0F  Could not check version for ${prefix}-engine, will attempt download`);
        binariesToProcess.push({
          prefix,
          needsUpdate: true,
          localVersion: null,
          remoteVersion: "latest"
        });
      }
    }
    const binariesToDownload = binariesToProcess.filter((b) => b.needsUpdate);
    if (binariesToDownload.length === 0) {
      this.mainSpinner.succeed(import_chalk.default.green("All binaries are up to date"));
      return;
    }
    this.mainSpinner.text = import_chalk.default.blue(`Updating ${binariesToDownload.length} binary(ies)...`);
    try {
      const results = await Promise.allSettled(binariesToDownload.map(async (binary) => {
        const maxRetries = 3;
        let attempt = 0;
        while (attempt <= maxRetries) {
          try {
            const binaryInfo = await this.getWithVersion(binary.prefix);
            if (!binaryInfo.name || !binaryInfo.url) {
              throw new Error(`Platform or architecture not supported for ${binary.prefix}`);
            }
            const tempZipPath = path.join(os2.tmpdir(), `dbcube-${binary.prefix}-${Date.now()}.zip`);
            const finalBinaryPath = path.join(binDir, binaryInfo.name);
            this.deleteOldBinary(binDir, binary.prefix);
            await this.downloadFileWithProgress(binaryInfo.url, tempZipPath, binary.prefix);
            await this.extractBinary(tempZipPath, finalBinaryPath, binary.prefix);
            console.log(`\u2705 ${binary.prefix}-engine updated to ${binary.remoteVersion}`);
            break;
          } catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Unknown error";
            if (attempt < maxRetries && (errorMessage.includes("ECONNRESET") || errorMessage.includes("timeout") || errorMessage.includes("ETIMEDOUT") || errorMessage.includes("ENOTFOUND"))) {
              attempt++;
              console.log(`\u{1F504} Retrying ${binary.prefix}-engine (${attempt}/${maxRetries})...`);
              await new Promise((resolve6) => setTimeout(resolve6, 1e3 + Math.random() * 1e3));
            } else {
              throw new Error(`Error downloading ${binary.prefix}: ${errorMessage}`);
            }
          }
        }
      }));
      const failures = results.filter((r) => r.status === "rejected");
      if (failures.length > 0) {
        const messages = failures.map((f) => f.reason instanceof Error ? f.reason.message : String(f.reason));
        throw new Error(messages.join(" | "));
      }
      this.mainSpinner.succeed(import_chalk.default.green("Binaries updated successfully"));
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      this.mainSpinner.fail(import_chalk.default.red(`Error updating binaries: ${errorMessage}`));
      throw error;
    }
  }
  static updateMainProgress(binary, current, total, status, attempt) {
    const progressBar = this.createProgressBar(current, total);
    const statusEmojis = {
      downloading: "\u{1F4E5}",
      extracting: "\u{1F4E6}",
      completed: "\u2705",
      exists: "\u2705",
      retrying: "\u{1F504}"
    };
    const statusMessages = {
      downloading: import_chalk.default.blue("descargando"),
      extracting: import_chalk.default.yellow("extrayendo"),
      completed: import_chalk.default.green("completado"),
      exists: import_chalk.default.gray("existe"),
      retrying: import_chalk.default.yellow(`reintentando (${attempt}/${3})`)
    };
    const emoji = statusEmojis[status] || "\u{1F4E5}";
    const message = statusMessages[status] || status;
    this.mainSpinner.text = `${progressBar} ${emoji} ${import_chalk.default.bold(binary)} - ${message}`;
  }
  static createProgressBar(current, total, width = 20) {
    const filled = Math.round(current / total * width);
    const empty = width - filled;
    const filledBar = import_chalk.default.green("\u2588".repeat(filled));
    const emptyBar = import_chalk.default.gray("\u2591".repeat(empty));
    const percentage = import_chalk.default.bold(`${current}/${total}`);
    return `[${filledBar}${emptyBar}] ${percentage}`;
  }
  static downloadFileWithProgress(url, outputPath, prefix) {
    return new Promise((resolve6, reject) => {
      const request = https.get(url, { timeout: 0 }, (response) => {
        if (response.statusCode === 302 || response.statusCode === 301) {
          const redirectUrl = response.headers.location;
          if (redirectUrl) {
            return this.downloadFileWithProgress(redirectUrl, outputPath, prefix).then(resolve6).catch(reject);
          }
        }
        if (response.statusCode !== 200) {
          reject(new Error(`HTTP ${response.statusCode} para ${prefix}`));
          return;
        }
        const file = fs.createWriteStream(outputPath);
        const totalBytes = parseInt(response.headers["content-length"] || "0", 10);
        let downloadedBytes = 0;
        response.on("data", (chunk) => {
          downloadedBytes += chunk.length;
          file.write(chunk);
          if (totalBytes > 0) {
            const progress = {
              downloaded: downloadedBytes,
              total: totalBytes,
              percentage: downloadedBytes / totalBytes * 100
            };
            this.updateDownloadProgress(prefix, progress);
          }
        });
        response.on("end", () => {
          file.end(() => resolve6());
        });
        response.on("error", (err) => {
          file.close();
          this.cleanupFile(outputPath);
          reject(err);
        });
        file.on("error", (err) => {
          file.close();
          this.cleanupFile(outputPath);
          reject(err);
        });
      });
      request.on("error", reject);
      request.on("timeout", () => {
        request.destroy();
        reject(new Error(`Timeout descargando ${prefix}`));
      });
    });
  }
  static updateDownloadProgress(binary, progress) {
    const percentage = progress.percentage.toFixed(1);
    const downloaded = (progress.downloaded / 1024 / 1024).toFixed(1);
    const total = (progress.total / 1024 / 1024).toFixed(1);
    const barWidth = 15;
    const filled = Math.round(progress.percentage / 100 * barWidth);
    const empty = barWidth - filled;
    const progressBar = import_chalk.default.green("\u2588".repeat(filled)) + import_chalk.default.gray("\u2591".repeat(empty));
    const progressText = `[${progressBar}] ${percentage}% (${downloaded}/${total}MB)`;
    this.mainSpinner.text = `\u{1F4E5} ${import_chalk.default.bold(binary)} - ${progressText}`;
  }
  static extractBinary(zipPath, outputPath, prefix) {
    return new Promise((resolve6, reject) => {
      const outDir = path.dirname(outputPath);
      let extracted = 0;
      const pending = [];
      fs.createReadStream(zipPath).pipe(unzipper.Parse()).on("entry", (entry) => {
        if (entry.type !== "File") {
          entry.autodrain();
          return;
        }
        const isMain = extracted === 0;
        extracted++;
        const target = isMain ? outputPath : path.join(outDir, path.basename(entry.path));
        pending.push(new Promise((res, rej) => {
          const writeStream = fs.createWriteStream(target);
          entry.pipe(writeStream);
          writeStream.on("finish", () => {
            if (process.platform !== "win32") {
              try {
                fs.chmodSync(target, 493);
              } catch {
              }
            }
            res();
          });
          writeStream.on("error", rej);
        }));
      }).on("error", (err) => {
        this.cleanupFile(zipPath);
        reject(err);
      }).on("close", () => {
        Promise.all(pending).then(() => {
          this.cleanupFile(zipPath);
          if (extracted === 0) {
            reject(new Error(`No se encontr\xF3 archivo v\xE1lido en el ZIP para ${prefix}`));
          } else {
            resolve6();
          }
        }).catch((err) => {
          this.cleanupFile(zipPath);
          reject(err);
        });
      });
    });
  }
  static cleanupFile(filePath) {
    try {
      if (fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
      }
    } catch {
    }
  }
  static getDefaultBinDir() {
    const __filename2 = typeof import_meta !== "undefined" && import_meta.url ? (0, import_url.fileURLToPath)(import_meta.url) : "";
    const __dirname = __filename2 ? (0, import_path.dirname)(__filename2) : process.cwd();
    const possibleDirs = [
      path.resolve(process.cwd(), ".dbcube", "bin"),
      path.resolve(process.cwd(), "node_modules", ".dbcube", "bin"),
      path.resolve(__dirname, "..", "bin")
    ];
    for (const dir of possibleDirs) {
      try {
        if (!fs.existsSync(dir)) {
          fs.mkdirSync(dir, { recursive: true });
        }
        const testFile = path.join(dir, ".test");
        fs.writeFileSync(testFile, "test");
        fs.unlinkSync(testFile);
        return dir;
      } catch {
        continue;
      }
    }
    const tempDir = path.join(os2.tmpdir(), ".dbcube", "bin");
    fs.mkdirSync(tempDir, { recursive: true });
    return tempDir;
  }
};

// src/lib/Binary.ts
var fs2 = __toESM(require("fs"));
var path2 = __toESM(require("path"));
var os3 = __toESM(require("os"));
var import_url2 = require("url");
var import_path2 = require("path");
var import_meta2 = {};
var Binary = class {
  static isDownloading = false;
  static downloadPromise = null;
  static cachedBinaries = {};
  static async ensureBinariesExist() {
    if (this.isDownloading && this.downloadPromise) {
      await this.downloadPromise;
      return;
    }
    const binDir = this.getBinDir();
    if (!this.isDownloading) {
      this.isDownloading = true;
      this.downloadPromise = this.downloadBinaries();
      try {
        await this.downloadPromise;
      } finally {
        this.isDownloading = false;
        this.downloadPromise = null;
      }
    }
  }
  static async downloadBinaries() {
    try {
      const binDir = this.getBinDir();
      await Downloader.download(binDir);
    } catch (error) {
      console.warn("\u26A0\uFE0F Dbcube: Error descargando binarios:", error.message);
      console.log("\u{1F527} Los binarios se intentar\xE1n descargar en la pr\xF3xima ejecuci\xF3n");
    }
  }
  static getBinDir() {
    const __filename2 = typeof import_meta2 !== "undefined" && import_meta2.url ? (0, import_url2.fileURLToPath)(import_meta2.url) : "";
    const __dirname = __filename2 ? (0, import_path2.dirname)(__filename2) : process.cwd();
    const possibleDirs = [
      path2.resolve(process.cwd(), ".dbcube", "bin"),
      path2.resolve(process.cwd(), "node_modules", ".dbcube", "bin"),
      path2.resolve(__dirname, "..", "bin")
    ];
    for (const dir of possibleDirs) {
      try {
        if (!fs2.existsSync(dir)) {
          fs2.mkdirSync(dir, { recursive: true });
        }
        return dir;
      } catch {
        continue;
      }
    }
    const tempDir = path2.join(os3.tmpdir(), ".dbcube", "bin");
    fs2.mkdirSync(tempDir, { recursive: true });
    return tempDir;
  }
  static findVersionedBinary(binDir, prefix, platform2, arch2) {
    const cacheKey = `${prefix}-${platform2}-${arch2}`;
    if (this.cachedBinaries[cacheKey]) {
      const cachedPath = path2.join(binDir, this.cachedBinaries[cacheKey]);
      if (fs2.existsSync(cachedPath)) {
        return cachedPath;
      }
    }
    try {
      const files = fs2.readdirSync(binDir);
      const extension = platform2 === "windows" ? ".exe" : "";
      const pattern = new RegExp(`^${prefix}-engine-v\\d+\\.\\d+\\.\\d+-${platform2}-${arch2}${extension.replace(".", "\\.")}$`);
      const matchingFile = files.find((f) => pattern.test(f));
      if (matchingFile) {
        this.cachedBinaries[cacheKey] = matchingFile;
        return path2.join(binDir, matchingFile);
      }
    } catch (error) {
    }
    const fallbackName = `${prefix}-engine-${platform2}-${arch2}${platform2 === "windows" ? ".exe" : ""}`;
    return path2.join(binDir, fallbackName);
  }
  static async get() {
    await this.ensureBinariesExist();
    const arch2 = new Arquitecture();
    const platform2 = arch2.getPlatform();
    const architecture = arch2.getArchitecture();
    const binDir = this.getBinDir();
    const platformMap = {
      windows: "windows",
      linux: "linux",
      darwin: "macos"
    };
    const archMap = {
      x86_64: "x64",
      aarch64: "arm64"
    };
    const plat = platformMap[platform2];
    const archSuffix = archMap[architecture];
    if (plat && archSuffix) {
      return {
        query_engine: this.findVersionedBinary(binDir, "query", plat, archSuffix),
        schema_engine: this.findVersionedBinary(binDir, "schema", plat, archSuffix)
      };
    }
    return {
      query_engine: "",
      schema_engine: ""
    };
  }
};

// src/lib/Config.ts
var Config = class {
  data = {};
  databases = {};
  /**
   * Establece la configuración
   * @param configData - Datos de configuración
   */
  set(configData) {
    this.data = configData;
    if (configData.databases) {
      this.databases = configData.databases;
    }
  }
  /**
   * Obtiene un valor de configuración
   * @param key - Clave de configuración
   */
  get(key) {
    return this.data[key];
  }
  /**
   * Obtiene la configuración de una base de datos específica
   * @param dbName - Nombre de la base de datos
   */
  getDatabase(dbName) {
    return this.databases[dbName] || null;
  }
  /**
   * Obtiene todas las bases de datos configuradas
   */
  getAllDatabases() {
    return this.databases;
  }
};

// src/lib/Engine.ts
var import_child_process2 = require("child_process");
var import_module = require("module");

// src/lib/DaemonClient.ts
var import_net = __toESM(require("net"));
var import_fs = __toESM(require("fs"));
var import_path3 = __toESM(require("path"));
var import_child_process = require("child_process");
var DaemonClient = class _DaemonClient {
  static registry = /* @__PURE__ */ new Map();
  name;
  binaryPath;
  engineArgs;
  socket = null;
  buffer = "";
  pending = [];
  starting = null;
  requestTimeout;
  constructor(name, binaryPath, engineArgs, requestTimeout = 3e4) {
    this.name = name;
    this.binaryPath = binaryPath;
    this.engineArgs = engineArgs;
    this.requestTimeout = requestTimeout;
  }
  static get(name, binaryPath, engineArgs) {
    let client = this.registry.get(name);
    if (!client) {
      client = new _DaemonClient(name, binaryPath, engineArgs);
      this.registry.set(name, client);
    }
    return client;
  }
  static isEnabled() {
    const flag = process.env.DBCUBE_DAEMON;
    if (flag === "0" || flag === "false") return false;
    return true;
  }
  portfilePath() {
    return import_path3.default.join(process.cwd(), ".dbcube", "daemon", `${this.name}.json`);
  }
  lockfilePath() {
    return import_path3.default.join(process.cwd(), ".dbcube", "daemon", `${this.name}.lock`);
  }
  /**
   * Exclusive-create lock so two processes never spawn two daemons for the
   * same database at once. Stale locks (crashed starter) expire after 15s.
   */
  acquireSpawnLock() {
    const lockPath = this.lockfilePath();
    try {
      import_fs.default.mkdirSync(import_path3.default.dirname(lockPath), { recursive: true });
      const fd = import_fs.default.openSync(lockPath, "wx");
      import_fs.default.writeSync(fd, String(process.pid));
      import_fs.default.closeSync(fd);
      return true;
    } catch {
      try {
        const age = Date.now() - import_fs.default.statSync(lockPath).mtimeMs;
        if (age > 15e3) {
          import_fs.default.unlinkSync(lockPath);
          return this.acquireSpawnLock();
        }
      } catch {
      }
      return false;
    }
  }
  releaseSpawnLock() {
    try {
      import_fs.default.unlinkSync(this.lockfilePath());
    } catch {
    }
  }
  /**
   * Ensures a usable daemon connection. Returns false when the daemon
   * can't be used (old binary, startup failure) so callers fall back
   * to one-shot spawn mode.
   */
  async ensure() {
    if (this.socket && !this.socket.destroyed) return true;
    if (this.starting) return this.starting;
    this.starting = this.connectOrStart().catch(() => false);
    const result = await this.starting;
    this.starting = null;
    return result;
  }
  async connectOrStart() {
    const port = this.readPortfile();
    if (port && await this.tryConnect(port)) return true;
    const gotLock = this.acquireSpawnLock();
    try {
      if (gotLock) {
        try {
          import_fs.default.unlinkSync(this.portfilePath());
        } catch {
        }
        this.spawnDaemon();
      }
      const deadline = Date.now() + 1e4;
      while (Date.now() < deadline) {
        await new Promise((r) => setTimeout(r, 150));
        const newPort = this.readPortfile();
        if (newPort && await this.tryConnect(newPort)) return true;
      }
      return false;
    } finally {
      if (gotLock) this.releaseSpawnLock();
    }
  }
  readPortfile() {
    try {
      const raw = import_fs.default.readFileSync(this.portfilePath(), "utf8");
      const info = JSON.parse(raw);
      const port = info?.port;
      return Number.isInteger(port) && port > 0 && port <= 65535 ? port : null;
    } catch {
      return null;
    }
  }
  spawnDaemon() {
    const child = (0, import_child_process.spawn)(this.binaryPath, [...this.engineArgs, "--action", "server", "--tcp-port", "9944"], {
      detached: true,
      stdio: "ignore",
      cwd: process.cwd()
    });
    child.unref();
  }
  tryConnect(port) {
    return new Promise((resolve6) => {
      const socket = import_net.default.createConnection({ host: "127.0.0.1", port }, async () => {
        socket.setNoDelay(true);
        this.attach(socket);
        try {
          const pong = await this.send({ action: "ping" }, 2e3);
          resolve6(pong.status === 200);
        } catch {
          this.detach();
          resolve6(false);
        }
      });
      socket.once("error", () => resolve6(false));
      socket.setTimeout(3e3, () => {
        socket.destroy();
        resolve6(false);
      });
    });
  }
  attach(socket) {
    this.socket = socket;
    this.buffer = "";
    socket.setTimeout(0);
    socket.on("data", (chunk) => this.onData(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)));
    socket.on("close", () => this.detach());
    socket.on("error", () => this.detach());
  }
  detach() {
    if (this.socket) {
      this.socket.removeAllListeners();
      this.socket.destroy();
      this.socket = null;
    }
    for (const req of this.pending.splice(0)) {
      clearTimeout(req.timer);
      req.reject(new Error("Daemon connection lost"));
    }
  }
  onData(chunk) {
    this.buffer += chunk.toString("utf8");
    let idx;
    while ((idx = this.buffer.indexOf("\n")) !== -1) {
      const line = this.buffer.slice(0, idx).trim();
      this.buffer = this.buffer.slice(idx + 1);
      if (!line) continue;
      const marker = line.indexOf("PROCESS_RESPONSE:");
      if (marker === -1) continue;
      const jsonPart = line.slice(marker + "PROCESS_RESPONSE:".length);
      const req = this.pending.shift();
      if (!req) continue;
      clearTimeout(req.timer);
      try {
        req.resolve(JSON.parse(jsonPart));
      } catch (e) {
        req.reject(new Error(`Invalid daemon response: ${e.message}`));
      }
    }
  }
  /**
   * Sends a request to the daemon. Responses arrive in FIFO order on the
   * single socket, so pending requests resolve in send order.
   */
  send(payload, timeoutMs) {
    return new Promise((resolve6, reject) => {
      if (!this.socket || this.socket.destroyed) {
        reject(new Error("Daemon not connected"));
        return;
      }
      const timer = setTimeout(() => {
        const i = this.pending.findIndex((p) => p.timer === timer);
        if (i !== -1) this.pending.splice(i, 1);
        reject(new Error("Daemon request timeout"));
        this.detach();
      }, timeoutMs ?? this.requestTimeout);
      this.pending.push({ resolve: resolve6, reject, timer });
      this.socket.write(JSON.stringify(payload) + "\n");
    });
  }
  async execute(dml, txId) {
    const payload = { action: "execute", dml };
    if (txId) payload.tx_id = txId;
    return this.send(payload);
  }
  async raw(query, params = [], txId) {
    const payload = { action: "raw", query, params };
    if (txId) payload.tx_id = txId;
    return this.send(payload);
  }
  async begin() {
    const res = await this.send({ action: "begin" });
    if (res.status !== 200 || !res.data?.tx_id) {
      throw new Error(res.message || "Failed to begin transaction");
    }
    return res.data.tx_id;
  }
  async commit(txId) {
    const res = await this.send({ action: "commit", tx_id: txId });
    if (res.status !== 200) throw new Error(res.message || "Failed to commit transaction");
  }
  async rollback(txId) {
    const res = await this.send({ action: "rollback", tx_id: txId });
    if (res.status !== 200) throw new Error(res.message || "Failed to rollback transaction");
  }
  async shutdown() {
    try {
      await this.send({ action: "shutdown" });
    } catch {
    }
    this.detach();
    try {
      import_fs.default.unlinkSync(this.portfilePath());
    } catch {
    }
  }
};

// src/lib/Engine.ts
var Engine = class {
  name;
  config;
  arguments;
  binary = null;
  timeout;
  daemonFailed = false;
  constructor(name, timeout = 3e4) {
    this.name = name;
    this.config = this.setConfig(name);
    this.arguments = this.setArguments();
    this.timeout = timeout;
  }
  async initializeBinary() {
    if (!this.binary) {
      this.binary = await Binary.get();
    }
  }
  /**
   * Returns a connected DaemonClient for this database, or null when
   * daemon mode is disabled/unavailable (then callers use spawn mode).
   */
  async getDaemon() {
    if (this.daemonFailed || !DaemonClient.isEnabled()) return null;
    await this.initializeBinary();
    if (!this.binary) return null;
    const binaryPath = this.binary["query_engine"];
    if (!binaryPath) return null;
    const client = DaemonClient.get(this.name, binaryPath, this.arguments);
    const ok2 = await client.ensure();
    if (!ok2) {
      this.daemonFailed = true;
      return null;
    }
    return client;
  }
  /**
   * Executes a DML plan. Uses the persistent daemon (sub-millisecond
   * overhead) when available; falls back to one-shot spawn otherwise.
   */
  async executeDml(dml, txId) {
    const daemon = await this.getDaemon();
    if (daemon) {
      try {
        return await daemon.execute(dml, txId);
      } catch (error) {
        if (txId) throw error;
      }
    }
    if (txId) {
      throw new Error("Transactions require daemon mode. Make sure the query-engine binary is up to date (npx dbcube update).");
    }
    return this.run("query_engine", ["--action", "execute", "--dml", JSON.stringify(dml)]);
  }
  /**
   * Executes raw SQL (or a MongoDB command document) with bound parameters.
   */
  async rawQuery(query, params = [], txId) {
    const daemon = await this.getDaemon();
    if (daemon) {
      try {
        return await daemon.raw(query, params, txId);
      } catch (error) {
        if (txId) throw error;
      }
    }
    if (txId) {
      throw new Error("Transactions require daemon mode. Make sure the query-engine binary is up to date (npx dbcube update).");
    }
    return this.run("query_engine", ["--action", "raw", "--query", query, "--params", JSON.stringify(params)]);
  }
  /** Starts a transaction (daemon mode only). Returns the transaction id. */
  async beginTransaction() {
    const daemon = await this.getDaemon();
    if (!daemon) {
      throw new Error("Transactions require daemon mode. Make sure the query-engine binary is up to date (npx dbcube update).");
    }
    return daemon.begin();
  }
  async commitTransaction(txId) {
    const daemon = await this.getDaemon();
    if (!daemon) throw new Error("Daemon connection lost during transaction");
    return daemon.commit(txId);
  }
  async rollbackTransaction(txId) {
    const daemon = await this.getDaemon();
    if (!daemon) throw new Error("Daemon connection lost during transaction");
    return daemon.rollback(txId);
  }
  setArguments() {
    let args = [];
    if (this.config.type == "sqlite") {
      args = [
        "--id",
        "dbcube-" + this.name,
        "--database-ref",
        this.name,
        "--database",
        this.config.config.DATABASE + ".db",
        "--motor",
        this.config.type
      ];
    } else {
      args = [
        "--id",
        "dbcube-" + this.name,
        "--database-ref",
        this.name,
        "--database",
        this.config.config.DATABASE,
        "--host",
        this.config.config.HOST,
        "--port",
        String(this.config.config.PORT),
        "--motor",
        this.config.type
      ];
      if (this.config.config.USER != null && this.config.config.USER !== "") {
        args.push("--user", this.config.config.USER);
      }
      if (this.config.config.PASSWORD != null && this.config.config.PASSWORD !== "") {
        args.push("--password", this.config.config.PASSWORD);
      }
    }
    return args;
  }
  setConfig(name) {
    const configInstance = new Config();
    try {
      const configFilePath = import_path4.default.resolve(process.cwd(), "dbcube.config.js");
      const requireUrl = typeof __filename !== "undefined" ? __filename : process.cwd();
      const require2 = (0, import_module.createRequire)(requireUrl);
      if (require2.cache && require2.resolve) {
        try {
          delete require2.cache[require2.resolve(configFilePath)];
        } catch (e) {
        }
      }
      const configModule = require2(configFilePath);
      const configFn = configModule.default || configModule;
      if (typeof configFn === "function") {
        configFn(configInstance);
      } else {
        console.error("\u274C El archivo dbcube.config.js no exporta una funci\xF3n.");
      }
    } catch (error) {
      console.error("\u274C Error loading config file:", error.message);
      if (error.code === "MODULE_NOT_FOUND") {
        console.error("\u274C Config file not found, please create a dbcube.config.js file");
      }
    }
    return configInstance.getDatabase(name);
  }
  getConfig() {
    return this.config;
  }
  async run(binary, args) {
    await this.initializeBinary();
    if (!this.binary) {
      throw new Error("Binary not initialized");
    }
    return new Promise((resolve6, reject) => {
      const child = (0, import_child_process2.spawn)(this.binary[binary], [...this.arguments, ...args]);
      let stdoutBuffer = "";
      let stderrBuffer = "";
      let isResolved = false;
      const timeoutId = setTimeout(() => {
        if (!isResolved) {
          isResolved = true;
          child.kill();
          reject(new Error("Process timeout"));
        }
      }, this.timeout);
      const resolveOnce = (response) => {
        if (!isResolved) {
          isResolved = true;
          clearTimeout(timeoutId);
          resolve6(response);
        }
      };
      const tryParseLines = (buffer) => {
        let idx;
        while ((idx = buffer.indexOf("\n")) !== -1) {
          const line = buffer.slice(0, idx);
          buffer = buffer.slice(idx + 1);
          const marker = line.indexOf("PROCESS_RESPONSE:");
          if (marker === -1) continue;
          try {
            const response = JSON.parse(line.slice(marker + "PROCESS_RESPONSE:".length));
            resolveOnce({
              status: response.status,
              message: response.message,
              data: response.data
            });
          } catch (error) {
            resolveOnce({
              status: 500,
              message: "Failed to parse response JSON",
              data: null
            });
          }
        }
        return buffer;
      };
      child.stdout.on("data", (data) => {
        const text = data.toString();
        const visible = text.split("\n").filter((l) => l.trim() && !l.includes("PROCESS_RESPONSE:")).join("\n");
        if (visible) console.log(visible);
        stdoutBuffer = tryParseLines(stdoutBuffer + text);
      });
      child.stderr.on("data", (data) => {
        const text = data.toString();
        const visible = text.split("\n").filter((l) => l.trim() && !l.includes("PROCESS_RESPONSE:")).join("\n");
        if (visible) console.log(visible);
        stderrBuffer = tryParseLines(stderrBuffer + text);
      });
      child.on("close", (code) => {
        clearTimeout(timeoutId);
        if (!isResolved) {
          resolveOnce({
            status: code === 0 ? 200 : 500,
            message: code === 0 ? "Process completed" : `Process exited with code ${code}`,
            data: null
          });
        }
      });
      child.on("error", (error) => {
        clearTimeout(timeoutId);
        if (!isResolved) {
          resolveOnce({
            status: 500,
            message: `Process error: ${error.message}`,
            data: null
          });
        }
      });
      child.unref();
    });
  }
};

// src/lib/QueryEngine.ts
var import_path5 = __toESM(require("path"));

// src/lib/EmbeddedEngine.ts
var fs4 = __toESM(require("fs"));
var path5 = __toESM(require("path"));
if (!process.env.UV_THREADPOOL_SIZE) {
  process.env.UV_THREADPOOL_SIZE = "16";
}
var backend = null;
var loadFailed = false;
var handles = /* @__PURE__ */ new Map();
var connecting = /* @__PURE__ */ new Map();
function platTag() {
  const arch2 = process.arch === "arm64" ? "arm64" : "x64";
  if (process.platform === "win32") return { plat: "windows", ext: "dll", arch: arch2 };
  if (process.platform === "darwin") return { plat: "macos", ext: "dylib", arch: arch2 };
  return { plat: "linux", ext: "so", arch: arch2 };
}
function findFile(name) {
  const candidates = [
    path5.resolve(process.cwd(), ".dbcube", "bin", name),
    path5.resolve(process.cwd(), "node_modules", ".dbcube", "bin", name)
  ];
  for (const c of candidates) {
    if (fs4.existsSync(c)) return c;
  }
  return null;
}
function ok(data) {
  return { status: 200, message: "OK", data };
}
function errResp(e) {
  const msg = e instanceof Error ? e.message : String(e);
  return { status: 500, message: msg, data: null };
}
function loadNapiBackend() {
  const { plat, arch: arch2 } = platTag();
  const addonPath = findFile(`query-engine-node-${plat}-${arch2}.node`);
  if (!addonPath) return null;
  try {
    const addon = require(addonPath);
    if (typeof addon.connect !== "function") return null;
    return {
      async connect(cfgJson) {
        return Number(await addon.connect(cfgJson));
      },
      async execute(handle, dml, txId) {
        try {
          const rows = await addon.execute(handle, dml, txId ?? void 0);
          return ok(rows);
        } catch (e) {
          return errResp(e);
        }
      },
      async raw(handle, query, params, txId) {
        try {
          const rows = await addon.raw(handle, query, params, txId ?? void 0);
          return ok(rows);
        } catch (e) {
          return errResp(e);
        }
      },
      async begin(handle) {
        try {
          const txId = await addon.begin(handle);
          return ok({ tx_id: txId });
        } catch (e) {
          return errResp(e);
        }
      },
      async commit(_handle, txId) {
        try {
          await addon.commit(txId);
          return ok(null);
        } catch (e) {
          return errResp(e);
        }
      },
      async rollback(_handle, txId) {
        try {
          await addon.rollback(txId);
          return ok(null);
        } catch (e) {
          return errResp(e);
        }
      },
      async disconnect(handle) {
        try {
          addon.disconnect(handle);
        } catch {
        }
      }
    };
  } catch {
    return null;
  }
}
function loadKoffiBackend() {
  const { plat, ext, arch: arch2 } = platTag();
  const libPath = findFile(`query-engine-embedded-${plat}-${arch2}.${ext}`);
  if (!libPath) return null;
  try {
    const koffi = require("koffi");
    const lib = koffi.load(libPath);
    const cConnect = lib.func("dbc_connect", "void *", ["str"]);
    const cExecute = lib.func("dbc_execute", "void *", ["uint64", "str", "str"]);
    const cRaw = lib.func("dbc_raw", "void *", ["uint64", "str", "str", "str"]);
    const cBegin = lib.func("dbc_begin", "void *", ["uint64"]);
    const cCommit = lib.func("dbc_commit", "void *", ["uint64", "str"]);
    const cRollback = lib.func("dbc_rollback", "void *", ["uint64", "str"]);
    const cDisconnect = lib.func("dbc_disconnect", "void *", ["uint64"]);
    const cFree = lib.func("dbc_free", "void", ["void *"]);
    const take = (ptr) => {
      if (!ptr) return { status: 500, message: "embedded engine returned null", data: null };
      const s = koffi.decode(ptr, "char", -1);
      cFree(ptr);
      try {
        const r = JSON.parse(s);
        return { status: r.status ?? 500, message: r.message ?? "", data: r.data ?? null };
      } catch (e) {
        return { status: 500, message: `Invalid embedded response: ${e.message}`, data: null };
      }
    };
    const call = (fn, ...args) => new Promise((resolve6, reject) => {
      fn.async(...args, (err, ptr) => {
        if (err) return reject(err);
        resolve6(take(ptr));
      });
    });
    return {
      async connect(cfgJson) {
        const res = await call(cConnect, cfgJson);
        if (res.status !== 200 || !res.data?.handle) {
          throw new Error(String(res.message || "embedded connect failed"));
        }
        return Number(res.data.handle);
      },
      execute: (handle, dml, txId) => call(cExecute, handle, JSON.stringify(dml), txId),
      raw: (handle, query, params, txId) => call(cRaw, handle, query, JSON.stringify(params ?? []), txId),
      begin: (handle) => call(cBegin, handle),
      commit: (handle, txId) => call(cCommit, handle, txId),
      rollback: (handle, txId) => call(cRollback, handle, txId),
      async disconnect(handle) {
        try {
          await call(cDisconnect, handle);
        } catch {
        }
      }
    };
  } catch {
    return null;
  }
}
function loadBackend() {
  if (backend) return backend;
  if (loadFailed) return null;
  if (process.env.DBCUBE_EMBEDDED === "0" || process.env.DBCUBE_EMBEDDED === "false") {
    loadFailed = true;
    return null;
  }
  backend = loadNapiBackend() ?? loadKoffiBackend();
  if (!backend) loadFailed = true;
  return backend;
}
var EmbeddedEngine = class {
  /** true si algún backend nativo está disponible en esta instalación. */
  static available() {
    return loadBackend() !== null;
  }
  static async handleFor(connectionId, config) {
    const existing = handles.get(connectionId);
    if (existing) return existing;
    const inFlight = connecting.get(connectionId);
    if (inFlight) return inFlight;
    const promise = (async () => {
      const be = loadBackend();
      if (!be) throw new Error("embedded engine not available");
      const motor = config.type === "postgres" ? "postgresql" : config.type;
      const cfg = {
        databaseRef: connectionId,
        motor,
        database: config.type === "sqlite" ? `${config.config.DATABASE}.db` : config.config.DATABASE,
        host: config.config.HOST,
        port: config.config.PORT != null ? Number(config.config.PORT) : void 0,
        user: config.config.USER,
        password: config.config.PASSWORD,
        maxConnections: config.pool?.maxConnections,
        minConnections: config.pool?.minConnections,
        acquireTimeoutMs: config.pool?.acquireTimeoutMs,
        idleTimeoutMs: config.pool?.idleTimeoutMs
      };
      const handle = await be.connect(JSON.stringify(cfg));
      handles.set(connectionId, handle);
      return handle;
    })();
    connecting.set(connectionId, promise);
    try {
      return await promise;
    } finally {
      connecting.delete(connectionId);
    }
  }
  static async executeDml(connectionId, config, dml, txId) {
    const be = loadBackend();
    const handle = await this.handleFor(connectionId, config);
    return be.execute(handle, dml, txId ?? null);
  }
  static async rawQuery(connectionId, config, query, params, txId) {
    const be = loadBackend();
    const handle = await this.handleFor(connectionId, config);
    return be.raw(handle, query, params ?? [], txId ?? null);
  }
  static async begin(connectionId, config) {
    const be = loadBackend();
    const handle = await this.handleFor(connectionId, config);
    return be.begin(handle);
  }
  static async commit(connectionId, config, txId) {
    const be = loadBackend();
    const handle = await this.handleFor(connectionId, config);
    return be.commit(handle, txId);
  }
  static async rollback(connectionId, config, txId) {
    const be = loadBackend();
    const handle = await this.handleFor(connectionId, config);
    return be.rollback(handle, txId);
  }
  static async disconnect(connectionId) {
    const be = loadBackend();
    const handle = handles.get(connectionId);
    if (be && handle) {
      handles.delete(connectionId);
      await be.disconnect(handle);
    }
  }
};

// src/lib/QueryEngine.ts
var import_module2 = require("module");
var net2 = __toESM(require("net"));
var import_child_process3 = require("child_process");
var globalTcpServers = /* @__PURE__ */ new Map();
var socketPools = /* @__PURE__ */ new Map();
var CLIENT_POOL_SIZE = 10;
var queryCache = /* @__PURE__ */ new Map();
var MAX_CACHE_SIZE = 500;
var QueryEngine = class {
  name;
  config;
  arguments;
  binary = null;
  timeout;
  connectionId;
  tcpPort;
  constructor(name, timeout = 3e4) {
    this.name = name;
    this.config = this.setConfig(name);
    this.arguments = this.setArguments();
    this.timeout = this.config?.daemon?.requestTimeoutMs ?? timeout;
    this.connectionId = `${name}_query_engine_${this.config.type}_${this.config.config.DATABASE}_${this.config.config.HOST || "localhost"}`;
    this.tcpPort = this.generatePort();
  }
  generatePort() {
    return 9944;
  }
  async findAvailablePort(startPort) {
    for (let port = startPort; port >= 9900; port--) {
      if (await this.isPortAvailable(port)) {
        return port;
      }
    }
    throw new Error("No available ports found in range 9900-9944");
  }
  isPortAvailable(port) {
    return new Promise((resolve6) => {
      const tester = net2.createServer();
      tester.once("error", () => resolve6(false));
      tester.once("listening", () => {
        tester.close();
        resolve6(true);
      });
      tester.listen(port, "127.0.0.1");
    });
  }
  async initializeBinary() {
    if (!this.binary) {
      this.binary = await Binary.get();
    }
  }
  setArguments() {
    let args = [];
    if (this.config.type == "sqlite") {
      args = [
        "--id",
        "dbcube-" + this.name,
        "--database-ref",
        this.name,
        "--database",
        this.config.config.DATABASE + ".db",
        "--motor",
        this.config.type
      ];
    } else {
      args = [
        "--id",
        "dbcube-" + this.name,
        "--database-ref",
        this.name,
        "--database",
        this.config.config.DATABASE,
        "--host",
        this.config.config.HOST,
        "--port",
        String(this.config.config.PORT),
        "--motor",
        this.config.type
      ];
      if (this.config.config.USER != null && this.config.config.USER !== "") {
        args.push("--user", this.config.config.USER);
      }
      if (this.config.config.PASSWORD != null && this.config.config.PASSWORD !== "") {
        args.push("--password", this.config.config.PASSWORD);
      }
    }
    const pool = this.config.pool ?? {};
    if (pool.maxConnections != null) args.push("--max-connections", String(pool.maxConnections));
    if (pool.minConnections != null) args.push("--min-connections", String(pool.minConnections));
    if (pool.acquireTimeoutMs != null) args.push("--acquire-timeout-ms", String(pool.acquireTimeoutMs));
    if (pool.idleTimeoutMs != null) args.push("--idle-timeout-ms", String(pool.idleTimeoutMs));
    return args;
  }
  /** El daemon puede desactivarse por config (daemon.enabled=false) o por env (DBCUBE_DAEMON=0). */
  daemonEnabled() {
    const flag = process.env.DBCUBE_DAEMON;
    if (flag === "0" || flag === "false") return false;
    return this.config?.daemon?.enabled !== false;
  }
  setConfig(name) {
    const configInstance = new Config();
    try {
      const configFilePath = import_path5.default.resolve(process.cwd(), "dbcube.config.js");
      const requireUrl = typeof __filename !== "undefined" ? __filename : process.cwd();
      const require2 = (0, import_module2.createRequire)(requireUrl);
      if (require2.cache && require2.resolve) {
        try {
          delete require2.cache[require2.resolve(configFilePath)];
        } catch (e) {
        }
      }
      const configModule = require2(configFilePath);
      const configFn = configModule.default || configModule;
      if (typeof configFn === "function") {
        configFn(configInstance);
      } else {
        console.error("\u274C El archivo dbcube.config.js no exporta una funci\xF3n.");
      }
    } catch (error) {
      console.error("\u274C Error loading config file:", error.message);
      if (error.code === "MODULE_NOT_FOUND") {
        console.error("\u274C Config file not found, please create a dbcube.config.js file");
      }
    }
    return configInstance.getDatabase(name);
  }
  getConfig() {
    return this.config;
  }
  async run(binary, args) {
    const actionIndex = args.findIndex((arg) => arg === "--action");
    const isExecuteAction = actionIndex !== -1 && args[actionIndex + 1] === "execute";
    const isQueryEngine = binary === "query_engine";
    if (isQueryEngine && isExecuteAction) {
      return this.executeWithTcpServer(this.argsToCommand(args));
    } else {
      return this.createProcess(binary, args);
    }
  }
  argsToCommand(args) {
    const command = {};
    for (let i = 0; i < args.length; i += 2) {
      if (args[i].startsWith("--")) {
        command[args[i].substring(2)] = args[i + 1];
      }
    }
    return command;
  }
  /** El motor embebido (FFI) es el camino por defecto cuando la librería
   *  nativa está disponible; el daemon TCP queda como fallback y como modo
   *  multi-proceso explícito. */
  useEmbedded() {
    return EmbeddedEngine.available();
  }
  /**
   * Executes a DML plan. Embedded engine when available (in-process FFI,
   * no network hop); persistent TCP daemon otherwise.
   * Pass txId to run it inside an active transaction.
   */
  async executeDml(dml, txId) {
    if (this.useEmbedded()) {
      return EmbeddedEngine.executeDml(this.connectionId, this.config, dml, txId);
    }
    const command = { action: "execute", dml };
    if (txId) command.tx_id = txId;
    return this.executeWithTcpServer(command);
  }
  /**
   * Executes raw SQL (or a MongoDB command document) with bound parameters.
   */
  async rawQuery(query, params = [], txId) {
    if (this.useEmbedded()) {
      return EmbeddedEngine.rawQuery(this.connectionId, this.config, query, params, txId);
    }
    const command = { action: "raw", query, params };
    if (txId) command.tx_id = txId;
    return this.executeWithTcpServer(command);
  }
  /** Starts a transaction and returns its id. */
  async beginTransaction() {
    const res = this.useEmbedded() ? await EmbeddedEngine.begin(this.connectionId, this.config) : await this.executeWithTcpServer({ action: "begin" });
    if (res.status !== 200 || !res.data?.tx_id) {
      throw new Error(String(res.message || "Failed to begin transaction (update the query-engine binary: npx dbcube update)"));
    }
    return res.data.tx_id;
  }
  async commitTransaction(txId) {
    const res = this.useEmbedded() ? await EmbeddedEngine.commit(this.connectionId, this.config, txId) : await this.executeWithTcpServer({ action: "commit", tx_id: txId });
    if (res.status !== 200) throw new Error(String(res.message || "Failed to commit transaction"));
  }
  async rollbackTransaction(txId) {
    const res = this.useEmbedded() ? await EmbeddedEngine.rollback(this.connectionId, this.config, txId) : await this.executeWithTcpServer({ action: "rollback", tx_id: txId });
    if (res.status !== 200) throw new Error(String(res.message || "Failed to rollback transaction"));
  }
  async executeWithTcpServer(command) {
    if (!this.daemonEnabled()) {
      if (command.tx_id || ["begin", "commit", "rollback"].includes(command.action)) {
        throw new Error("Transactions require daemon mode. Remove daemon.enabled=false from dbcube.config.js (or unset DBCUBE_DAEMON=0).");
      }
      const args = [];
      for (const [key, value] of Object.entries(command)) {
        args.push(`--${key.replace(/_/g, "-")}`, typeof value === "string" ? value : JSON.stringify(value));
      }
      return this.createProcess("query_engine", args);
    }
    const serverInfo = globalTcpServers.get(this.connectionId);
    if (!serverInfo || !serverInfo.process || serverInfo.process.killed) {
      await this.startTcpServer();
      await this.waitForServerReady();
    }
    const port = globalTcpServers.get(this.connectionId).port;
    const socket = await this.acquireSocket(port);
    try {
      const result = await this.executeOnConnection(socket, command);
      this.releaseSocket(socket, false);
      return result;
    } catch (error) {
      this.releaseSocket(socket, true);
      throw error;
    }
  }
  poolFor() {
    let pool = socketPools.get(this.connectionId);
    if (!pool) {
      pool = { free: [], total: 0, waiters: [] };
      socketPools.set(this.connectionId, pool);
    }
    return pool;
  }
  async acquireSocket(port) {
    const pool = this.poolFor();
    while (pool.free.length > 0) {
      const s = pool.free.pop();
      if (!s.destroyed) return s;
      pool.total--;
    }
    if (pool.total < CLIENT_POOL_SIZE) {
      pool.total++;
      try {
        return await this.createNewConnection(port);
      } catch (e) {
        pool.total--;
        throw e;
      }
    }
    return new Promise((resolve6) => pool.waiters.push(resolve6));
  }
  releaseSocket(socket, broken) {
    const pool = socketPools.get(this.connectionId);
    if (!pool) {
      socket.destroy();
      return;
    }
    if (broken || socket.destroyed) {
      pool.total--;
      try {
        socket.destroy();
      } catch {
      }
      return;
    }
    const waiter = pool.waiters.shift();
    if (waiter) {
      waiter(socket);
    } else {
      pool.free.push(socket);
    }
  }
  async waitForServerReady() {
    const maxRetries = 10;
    const retryDelay = 500;
    const serverInfo = globalTcpServers.get(this.connectionId);
    if (!serverInfo) {
      throw new Error("Server not found");
    }
    for (let i = 0; i < maxRetries; i++) {
      if (await this.isServerResponding(serverInfo.port)) {
        return;
      }
      await this.sleep(retryDelay);
    }
    throw new Error("TCP server failed to become ready within timeout");
  }
  async isServerResponding(port) {
    return new Promise((resolve6) => {
      const client = new net2.Socket();
      const timeout = setTimeout(() => {
        client.destroy();
        resolve6(false);
      }, 1e3);
      client.connect(port, "127.0.0.1", () => {
        clearTimeout(timeout);
        client.destroy();
        resolve6(true);
      });
      client.on("error", () => {
        clearTimeout(timeout);
        resolve6(false);
      });
    });
  }
  sleep(ms) {
    return new Promise((resolve6) => setTimeout(resolve6, ms));
  }
  getCachedDML(dmlJson) {
    if (queryCache.has(dmlJson)) {
      return queryCache.get(dmlJson);
    }
    const parsed = JSON.parse(dmlJson);
    if (queryCache.size >= MAX_CACHE_SIZE) {
      const oldest = queryCache.keys().next().value;
      if (oldest !== void 0) queryCache.delete(oldest);
    }
    queryCache.set(dmlJson, parsed);
    return parsed;
  }
  async startTcpServer() {
    await this.initializeBinary();
    if (!this.binary) {
      throw new Error("Binary not initialized");
    }
    this.tcpPort = await this.findAvailablePort(this.tcpPort);
    return new Promise((resolve6, reject) => {
      const serverArgs = [...this.arguments, "--action", "server", "--tcp-port", this.tcpPort.toString()];
      const serverProcess = (0, import_child_process3.spawn)(this.binary["query_engine"], serverArgs);
      let started = false;
      const timeout = setTimeout(() => {
        if (!started) {
          serverProcess.kill();
          reject(new Error("TCP server startup timeout"));
        }
      }, 15e3);
      serverProcess.stdout.on("data", (data) => {
        const output = data.toString();
        if (output.includes("TCP Server listening on")) {
          if (!started) {
            started = true;
            clearTimeout(timeout);
            globalTcpServers.set(this.connectionId, {
              port: this.tcpPort,
              process: serverProcess
            });
            resolve6();
          }
        }
      });
      serverProcess.stderr.on("data", (data) => {
        console.error(`[query_engine TCP Server Error] ${data.toString().trim()}`);
      });
      serverProcess.on("close", (code) => {
        globalTcpServers.delete(this.connectionId);
      });
      serverProcess.on("error", (error) => {
        if (!started) {
          clearTimeout(timeout);
          reject(error);
        }
      });
    });
  }
  async createNewConnection(port) {
    return new Promise((resolve6, reject) => {
      const client = new net2.Socket();
      client.setNoDelay(true);
      client.setKeepAlive(true, 6e4);
      const timeout = setTimeout(() => {
        client.destroy();
        reject(new Error("Connection timeout"));
      }, 5e3);
      client.connect(port, "127.0.0.1", () => {
        clearTimeout(timeout);
        resolve6(client);
      });
      client.on("error", (error) => {
        clearTimeout(timeout);
        reject(error);
      });
    });
  }
  async executeOnConnection(connection, command) {
    return new Promise((resolve6, reject) => {
      let responseBuffer = "";
      let isResolved = false;
      const timeout = setTimeout(() => {
        if (!isResolved) {
          isResolved = true;
          connection.removeListener("data", onData);
          connection.removeListener("error", onError);
          reject(new Error("Request timeout"));
        }
      }, this.timeout);
      const onData = (data) => {
        responseBuffer += data.toString();
        let idx;
        while ((idx = responseBuffer.indexOf("\n")) !== -1 && !isResolved) {
          const line = responseBuffer.slice(0, idx);
          responseBuffer = responseBuffer.slice(idx + 1);
          const marker = line.indexOf("PROCESS_RESPONSE:");
          if (marker === -1) continue;
          isResolved = true;
          clearTimeout(timeout);
          connection.removeListener("data", onData);
          connection.removeListener("error", onError);
          try {
            const response = JSON.parse(line.slice(marker + "PROCESS_RESPONSE:".length));
            resolve6({
              status: response.status,
              message: response.message,
              data: response.data
            });
          } catch (error) {
            reject(new Error("Failed to parse response"));
          }
        }
      };
      const onError = (error) => {
        if (!isResolved) {
          isResolved = true;
          clearTimeout(timeout);
          connection.removeListener("data", onData);
          connection.removeListener("error", onError);
          reject(error);
        }
      };
      connection.on("data", onData);
      connection.on("error", onError);
      connection.write(JSON.stringify(command) + "\n");
    });
  }
  async createProcess(binary, args) {
    await this.initializeBinary();
    if (!this.binary) {
      throw new Error("Binary not initialized");
    }
    return new Promise((resolve6, reject) => {
      const child = (0, import_child_process3.spawn)(this.binary[binary], [...this.arguments, ...args]);
      let stdoutBuffer = "";
      let stderrBuffer = "";
      let isResolved = false;
      const timeoutId = setTimeout(() => {
        if (!isResolved) {
          isResolved = true;
          child.kill();
          reject(new Error("Process timeout"));
        }
      }, this.timeout);
      const resolveOnce = (response) => {
        if (!isResolved) {
          isResolved = true;
          clearTimeout(timeoutId);
          resolve6(response);
        }
      };
      const tryParseLines = (buffer) => {
        let idx;
        while ((idx = buffer.indexOf("\n")) !== -1) {
          const line = buffer.slice(0, idx);
          buffer = buffer.slice(idx + 1);
          const marker = line.indexOf("PROCESS_RESPONSE:");
          if (marker === -1) continue;
          try {
            const response = JSON.parse(line.slice(marker + "PROCESS_RESPONSE:".length));
            resolveOnce({
              status: response.status,
              message: response.message,
              data: response.data
            });
          } catch (error) {
            resolveOnce({
              status: 500,
              message: "Failed to parse response JSON",
              data: null
            });
          }
        }
        return buffer;
      };
      child.stdout.on("data", (data) => {
        stdoutBuffer = tryParseLines(stdoutBuffer + data.toString());
      });
      child.stderr.on("data", (data) => {
        stderrBuffer = tryParseLines(stderrBuffer + data.toString());
      });
      child.on("close", (code) => {
        clearTimeout(timeoutId);
        if (!isResolved) {
          resolveOnce({
            status: code === 0 ? 200 : 500,
            message: code === 0 ? "Process completed" : `Process exited with code ${code}`,
            data: null
          });
        }
      });
      child.on("error", (error) => {
        clearTimeout(timeoutId);
        if (!isResolved) {
          resolveOnce({
            status: 500,
            message: `Process error: ${error.message}`,
            data: null
          });
        }
      });
      child.unref();
    });
  }
  async disconnect() {
    await EmbeddedEngine.disconnect(this.connectionId);
    const pool = socketPools.get(this.connectionId);
    if (pool) {
      for (const s of pool.free) {
        try {
          s.destroy();
        } catch {
        }
      }
      socketPools.delete(this.connectionId);
    }
    const serverInfo = globalTcpServers.get(this.connectionId);
    if (serverInfo && serverInfo.process && !serverInfo.process.killed) {
      serverInfo.process.kill();
      globalTcpServers.delete(this.connectionId);
      return {
        status: 200,
        message: "TCP server and connections stopped",
        data: null
      };
    }
    return {
      status: 200,
      message: "No TCP server to stop",
      data: null
    };
  }
};

// src/lib/SqliteExecutor.ts
var import_child_process4 = require("child_process");
var path7 = __toESM(require("path"));
var fs5 = __toESM(require("fs"));
var import_util = require("util");
var import_module3 = require("module");
var import_url3 = require("url");
var import_path6 = require("path");
var import_meta3 = {};
var execAsync = (0, import_util.promisify)(import_child_process4.exec);
var SqliteExecutor = class {
  binaryPath;
  dbPath;
  constructor(dbPath) {
    this.dbPath = dbPath;
    this.binaryPath = this.getBinaryPath();
  }
  findVersionedBinary(binDir, platform2) {
    try {
      const files = fs5.readdirSync(binDir);
      const extension = platform2 === "win32" ? ".exe" : "";
      const platformName = platform2 === "win32" ? "windows" : platform2 === "darwin" ? "macos" : "linux";
      const pattern = new RegExp(`^sqlite-engine-v\\d+\\.\\d+\\.\\d+-${platformName}-x64${extension.replace(".", "\\.")}$`);
      const matchingFile = files.find((f) => pattern.test(f));
      if (matchingFile) {
        return path7.join(binDir, matchingFile);
      }
    } catch (error) {
    }
    return null;
  }
  getBinaryPath() {
    const __filename2 = typeof import_meta3 !== "undefined" && import_meta3.url ? (0, import_url3.fileURLToPath)(import_meta3.url) : "";
    const __dirname = __filename2 ? (0, import_path6.dirname)(__filename2) : process.cwd();
    const possibleDirs = [
      path7.resolve(process.cwd(), ".dbcube", "bin"),
      path7.resolve(process.cwd(), "node_modules", ".dbcube", "bin"),
      path7.resolve(__dirname, "..", "bin")
    ];
    const platform2 = process.platform;
    const extension = platform2 === "win32" ? ".exe" : "";
    const platformName = platform2 === "win32" ? "windows" : platform2 === "darwin" ? "macos" : "linux";
    for (const dir of possibleDirs) {
      const versionedPath = this.findVersionedBinary(dir, platform2);
      if (versionedPath && fs5.existsSync(versionedPath)) {
        return versionedPath;
      }
    }
    const binaryName = `sqlite-engine-${platformName}-x64${extension}`;
    for (const dir of possibleDirs) {
      const fullPath = path7.join(dir, binaryName);
      if (fs5.existsSync(fullPath)) {
        return fullPath;
      }
    }
    const fallbackName = `sqlite-engine${extension}`;
    for (const dir of possibleDirs) {
      const fullPath = path7.join(dir, fallbackName);
      if (fs5.existsSync(fullPath)) {
        return fullPath;
      }
    }
    return path7.join(possibleDirs[0], binaryName);
  }
  async executeBinary(args) {
    const escapedArgs = args.map((arg) => {
      if (arg.includes(" ") || arg.includes('"') || arg.includes("(") || arg.includes(")")) {
        return `"${arg.replace(/"/g, '\\"')}"`;
      }
      return arg;
    });
    const command = `"${this.binaryPath}" ${escapedArgs.join(" ")}`;
    try {
      const { stdout, stderr } = await execAsync(command, {
        maxBuffer: 10 * 1024 * 1024,
        // 10MB buffer
        timeout: 3e4
        // 30s timeout
      });
      if (stderr && stderr.trim()) {
        console.warn("SQLite Engine Warning:", stderr);
      }
      const result = JSON.parse(stdout.trim());
      return result;
    } catch (error) {
      if (error.stdout) {
        try {
          const result = JSON.parse(error.stdout.trim());
          return result;
        } catch (parseError) {
        }
      }
      throw new Error(`SQLite execution failed: ${error.message}`);
    }
  }
  async connect() {
    try {
      const result = await this.executeBinary([
        "--action",
        "connect",
        "--database",
        this.dbPath
      ]);
      return result.status === "success";
    } catch (error) {
      return false;
    }
  }
  async exists() {
    try {
      const result = await this.executeBinary([
        "--action",
        "exists",
        "--database",
        this.dbPath
      ]);
      return result.status === "success" && result.data === true;
    } catch (error) {
      return false;
    }
  }
  async query(sql, params) {
    const args = [
      "--action",
      "query",
      "--database",
      this.dbPath,
      "--query",
      sql
    ];
    if (params && params.length > 0) {
      args.push("--params", JSON.stringify(params));
    }
    return this.executeBinary(args);
  }
  // Método para múltiples queries (como lo hace better-sqlite3)
  async queryMultiple(sql) {
    return this.query(sql);
  }
  // Método prepare que simula prepared statements
  prepare(sql) {
    return {
      all: async (...params) => {
        const result = await this.query(sql, params);
        if (result.status === "error") {
          throw new Error(result.message);
        }
        return Array.isArray(result.data) ? result.data : [];
      },
      run: async (...params) => {
        const result = await this.query(sql, params);
        if (result.status === "error") {
          throw new Error(result.message);
        }
        if (typeof result.data === "object" && result.data.hasOwnProperty("changes") && result.data.hasOwnProperty("lastID")) {
          return result.data;
        }
        return { changes: 0, lastID: 0 };
      }
    };
  }
  // Para compatibilidad con better-sqlite3 API sincrona usando deasync si es necesario
  prepareSync(sql) {
    const __filename2 = typeof import_meta3 !== "undefined" && import_meta3.url ? (0, import_url3.fileURLToPath)(import_meta3.url) : "";
    const requireUrl = __filename2 || import_meta3.url;
    const require2 = (0, import_module3.createRequire)(requireUrl);
    const deasync = require2("deasync");
    return {
      all: (...params) => {
        let result;
        let done = false;
        let error;
        this.query(sql, params).then((res) => {
          if (res.status === "error") {
            error = new Error(res.message);
          } else {
            result = Array.isArray(res.data) ? res.data : [];
          }
          done = true;
        }).catch((err) => {
          error = err;
          done = true;
        });
        deasync.loopWhile(() => !done);
        if (error) throw error;
        return result;
      },
      run: (...params) => {
        let result;
        let done = false;
        let error;
        this.query(sql, params).then((res) => {
          if (res.status === "error") {
            error = new Error(res.message);
          } else if (typeof res.data === "object" && res.data.hasOwnProperty("changes") && res.data.hasOwnProperty("lastID")) {
            result = res.data;
          } else {
            result = { changes: 0, lastID: 0 };
          }
          done = true;
        }).catch((err) => {
          error = err;
          done = true;
        });
        deasync.loopWhile(() => !done);
        if (error) throw error;
        return result;
      }
    };
  }
};

// src/lib/DbConfig.ts
var path8 = __toESM(require("path"));
var import_fs2 = __toESM(require("fs"));
var rootPath = path8.resolve(process.cwd(), ".dbcube");
var SQLite = class {
  executor = null;
  database;
  constructor(config) {
    this.database = config.DATABASE;
  }
  async ifExist() {
    if (this.database) {
      const dbPath = this.database || ":memory:";
      const configPath = path8.join(rootPath, dbPath + ".db");
      if (!import_fs2.default.existsSync(rootPath)) {
        import_fs2.default.mkdirSync(rootPath, { recursive: true });
      }
      if (import_fs2.default.existsSync(configPath)) {
        return true;
      }
      if (!this.executor) {
        this.executor = new SqliteExecutor(configPath);
      }
      return await this.executor.exists();
    }
    return false;
  }
  async connect() {
    return new Promise(async (resolve6, reject) => {
      try {
        if (!this.executor) {
          const dbPath = this.database || ":memory:";
          const configPath = path8.join(rootPath, dbPath + ".db");
          if (!import_fs2.default.existsSync(rootPath)) {
            import_fs2.default.mkdirSync(rootPath, { recursive: true });
          }
          this.executor = new SqliteExecutor(configPath);
          const connected = await this.executor.connect();
          if (!connected) {
            throw new Error("Failed to connect to SQLite database");
          }
        }
        resolve6(this.executor);
      } catch (error) {
        reject(error);
      }
    });
  }
  async disconnect() {
    return new Promise((resolve6) => {
      if (this.executor) {
        this.executor = null;
      }
      resolve6();
    });
  }
  async query(sqlQuery) {
    return new Promise(async (resolve6) => {
      try {
        if (typeof sqlQuery !== "string") {
          throw new Error("The SQL query must be a string.");
        }
        if (!this.executor) {
          await this.connect();
        }
        if (!this.executor) {
          throw new Error("Database connection is not available.");
        }
        const result = await this.executor.queryMultiple(sqlQuery);
        if (result.status === "error") {
          resolve6({
            status: "error",
            message: result.message,
            data: null
          });
        } else {
          resolve6({
            status: "success",
            message: "Query executed successfully",
            data: result.data
          });
        }
      } catch (error) {
        resolve6({
          status: "error",
          message: error.message || "An error occurred while executing the query.",
          data: null
        });
      }
    });
  }
  async queryWithParameters(sqlQuery, params = []) {
    return new Promise(async (resolve6) => {
      try {
        if (typeof sqlQuery !== "string") {
          throw new Error("The SQL query must be a string.");
        }
        if (!Array.isArray(params)) {
          throw new Error("Parameters must be an array.");
        }
        if (!this.executor) {
          await this.connect();
        }
        if (!this.executor) {
          throw new Error("Database connection is not available.");
        }
        const result = await this.executor.query(sqlQuery, params);
        if (result.status === "error") {
          resolve6({
            status: "error",
            message: result.message,
            data: null
          });
        } else {
          resolve6({
            status: "success",
            message: "Query executed successfully",
            data: result.data
          });
        }
      } catch (error) {
        console.log(error);
        resolve6({
          status: "error",
          message: error.message || "An error occurred while executing the query.",
          data: null
        });
      }
    });
  }
  convertToParameterizedQuery(sql) {
    const normalizedSql = sql.replace(/\s+/g, " ").trim();
    const baseQueryMatch = normalizedSql.match(/^(.+?)\s+VALUES\s*\(/i);
    if (!baseQueryMatch) {
      throw new Error("No se pudo encontrar la estructura VALUES en la consulta");
    }
    const baseQuery = baseQueryMatch[1];
    const valuesStartIndex = normalizedSql.toUpperCase().indexOf("VALUES");
    const valuesSection = normalizedSql.substring(valuesStartIndex);
    const valuesMatch = valuesSection.match(/VALUES\s*\((.+)\)\s*;?\s*$/i);
    if (!valuesMatch) {
      throw new Error("No se pudieron extraer los valores de la consulta");
    }
    const valuesString = valuesMatch[1];
    function parseValues(str) {
      const values = [];
      let currentValue = "";
      let inQuotes = false;
      let quoteChar = "";
      let inCompute = false;
      let computeDepth = 0;
      for (let i = 0; i < str.length; i++) {
        const char = str[i];
        const prevChar = str[i - 1];
        if (!inQuotes && str.substring(i, i + 8) === "@compute") {
          inCompute = true;
        }
        if (inCompute && char === "(") {
          computeDepth++;
        } else if (inCompute && char === ")") {
          computeDepth--;
          if (computeDepth === 0) {
            inCompute = false;
          }
        }
        if (!inCompute && (char === '"' || char === "'") && prevChar !== "\\") {
          if (!inQuotes) {
            inQuotes = true;
            quoteChar = char;
          } else if (char === quoteChar) {
            if (str[i + 1] === quoteChar) {
              currentValue += char + char;
              i++;
              continue;
            } else {
              inQuotes = false;
              quoteChar = "";
            }
          }
        }
        if (!inQuotes && !inCompute && char === ",") {
          values.push(cleanValue(currentValue.trim()));
          currentValue = "";
          continue;
        }
        currentValue += char;
      }
      if (currentValue.trim()) {
        values.push(cleanValue(currentValue.trim()));
      }
      return values;
    }
    function cleanValue(value) {
      value = value.trim();
      if (value.startsWith("@compute")) {
        return value;
      }
      if (value.startsWith('"') && value.endsWith('"') || value.startsWith("'") && value.endsWith("'")) {
        value = value.slice(1, -1);
        value = value.replace(/''/g, "'").replace(/""/g, '"');
      }
      return value;
    }
    const parameters = parseValues(valuesString);
    const placeholders = parameters.map(() => "?").join(", ");
    const parametrizedQuery = `${baseQuery} VALUES (${placeholders});`;
    return {
      query: parametrizedQuery,
      parameters
    };
  }
};
var DbConfig = new SQLite({ DATABASE: "config" });
var DbConfig_default = DbConfig;

// src/lib/FileLogger.ts
var fs7 = __toESM(require("fs"));
var path9 = __toESM(require("path"));
var import_events = require("events");
var FileLogger = class _FileLogger extends import_events.EventEmitter {
  static watchers = /* @__PURE__ */ new Map();
  // Store listener functions
  static buffers = /* @__PURE__ */ new Map();
  /**
   * Escribe un log en el archivo especificado
   * @param filePath - Ruta del archivo de log
   * @param message - Mensaje a escribir
   * @param level - Nivel del log (INFO, ERROR, WARN, DEBUG)
   * @param append - Si debe agregar al final del archivo (default: true)
   */
  static async write(filePath, message, level = "INFO", append = true) {
    try {
      const dir = path9.dirname(filePath);
      if (!fs7.existsSync(dir)) {
        fs7.mkdirSync(dir, { recursive: true });
      }
      const timestamp = (/* @__PURE__ */ new Date()).toISOString();
      const formattedMessage = `[${timestamp}] [${level}] ${message}
`;
      if (_FileLogger.buffers.has(filePath)) {
        _FileLogger.buffers.get(filePath).push(formattedMessage);
        return true;
      }
      if (append) {
        await fs7.promises.appendFile(filePath, formattedMessage, "utf8");
      } else {
        await fs7.promises.writeFile(filePath, formattedMessage, "utf8");
      }
      return true;
    } catch (error) {
      console.error("Error escribiendo log:", error);
      throw error;
    }
  }
  /**
   * Inicia un buffer temporal para un archivo de log
   * @param filePath - Ruta del archivo de log
   */
  static startBuffer(filePath) {
    if (!_FileLogger.buffers.has(filePath)) {
      _FileLogger.buffers.set(filePath, []);
    }
  }
  /**
   * Confirma y escribe todos los logs del buffer al archivo
   * @param filePath - Ruta del archivo de log
   */
  static async commitBuffer(filePath) {
    const buffer = _FileLogger.buffers.get(filePath);
    if (buffer && buffer.length > 0) {
      try {
        const dir = path9.dirname(filePath);
        if (!fs7.existsSync(dir)) {
          fs7.mkdirSync(dir, { recursive: true });
        }
        const content = buffer.join("");
        await fs7.promises.appendFile(filePath, content, "utf8");
        _FileLogger.buffers.delete(filePath);
        return true;
      } catch (error) {
        console.error("Error confirmando buffer:", error);
        throw error;
      }
    }
    _FileLogger.buffers.delete(filePath);
    return false;
  }
  /**
   * Descarta todos los logs del buffer sin escribirlos
   * @param filePath - Ruta del archivo de log
   */
  static discardBuffer(filePath) {
    if (_FileLogger.buffers.has(filePath)) {
      const discardedCount = _FileLogger.buffers.get(filePath).length;
      _FileLogger.buffers.delete(filePath);
      return discardedCount;
    }
    return 0;
  }
  /**
   * Verifica si existe un buffer activo para un archivo
   * @param filePath - Ruta del archivo de log
   */
  static hasBuffer(filePath) {
    return _FileLogger.buffers.has(filePath);
  }
  /**
   * Obtiene el número de logs en el buffer
   * @param filePath - Ruta del archivo de log
   */
  static getBufferSize(filePath) {
    const buffer = _FileLogger.buffers.get(filePath);
    return buffer ? buffer.length : 0;
  }
  /**
   * Intercepta console.log y console.error para escribir a archivo
   * @param filePath - Ruta del archivo de log
   * @param options - Opciones de interceptación
   */
  static interceptConsole(filePath, options = {}) {
    const {
      interceptLog = true,
      interceptError = true,
      interceptWarn = true,
      keepOriginal = true,
      // Mantener el comportamiento original de console
      useBuffer = false
      // Usar buffer temporal
    } = options;
    if (useBuffer) {
      _FileLogger.startBuffer(filePath);
    }
    const originalLog = console.log;
    const originalError = console.error;
    const originalWarn = console.warn;
    if (interceptLog) {
      console.log = function(...args) {
        const message = args.map(
          (arg) => typeof arg === "object" ? JSON.stringify(arg) : String(arg)
        ).join(" ");
        _FileLogger.write(filePath, message, "INFO").catch((err) => {
          originalError("Error escribiendo log:", err);
        });
        if (keepOriginal) {
          originalLog.apply(console, args);
        }
      };
    }
    if (interceptError) {
      console.error = function(...args) {
        const message = args.map(
          (arg) => typeof arg === "object" ? JSON.stringify(arg) : String(arg)
        ).join(" ");
        _FileLogger.write(filePath, message, "ERROR").catch((err) => {
          originalError("Error escribiendo error log:", err);
        });
        if (keepOriginal) {
          originalError.apply(console, args);
        }
      };
    }
    if (interceptWarn) {
      console.warn = function(...args) {
        const message = args.map(
          (arg) => typeof arg === "object" ? JSON.stringify(arg) : String(arg)
        ).join(" ");
        _FileLogger.write(filePath, message, "WARN").catch((err) => {
          originalError("Error escribiendo warn log:", err);
        });
        if (keepOriginal) {
          originalWarn.apply(console, args);
        }
      };
    }
    return {
      restore: () => {
        if (interceptLog) console.log = originalLog;
        if (interceptError) console.error = originalError;
        if (interceptWarn) console.warn = originalWarn;
      },
      commit: async () => {
        if (useBuffer) {
          return await _FileLogger.commitBuffer(filePath);
        }
        return false;
      },
      discard: () => {
        if (useBuffer) {
          return _FileLogger.discardBuffer(filePath);
        }
        return 0;
      },
      hasBuffer: () => _FileLogger.hasBuffer(filePath),
      getBufferSize: () => _FileLogger.getBufferSize(filePath)
    };
  }
  /**
   * Lee el contenido completo del archivo de log
   * @param filePath - Ruta del archivo de log
   * @param options - Opciones de lectura
   * @returns Contenido del archivo
   */
  static async read(filePath, options = {}) {
    const {
      lines = null,
      // Número de líneas a leer (null = todas)
      fromEnd = false,
      // Si debe leer desde el final
      asArray = false
      // Si debe retornar como array de líneas
    } = options;
    try {
      if (!fs7.existsSync(filePath)) {
        return asArray ? [] : "";
      }
      let content = await fs7.promises.readFile(filePath, "utf8");
      if (asArray) {
        let linesArray = content.split("\n").filter((line) => line.trim() !== "");
        if (lines !== null) {
          if (fromEnd) {
            linesArray = linesArray.slice(-lines);
          } else {
            linesArray = linesArray.slice(0, lines);
          }
        }
        return linesArray;
      }
      if (lines !== null) {
        let linesArray = content.split("\n").filter((line) => line.trim() !== "");
        if (fromEnd) {
          linesArray = linesArray.slice(-lines);
        } else {
          linesArray = linesArray.slice(0, lines);
        }
        content = linesArray.join("\n");
      }
      return content;
    } catch (error) {
      console.error("Error leyendo log:", error);
      throw error;
    }
  }
  /**
   * Observa un archivo de log en tiempo real
   * @param filePath - Ruta del archivo de log
   * @param callback - Función callback para nuevas líneas
   * @param options - Opciones del watcher
   * @returns Objeto con métodos para controlar el watcher
   */
  static watch(filePath, callback, options = {}) {
    const {
      persistent = true,
      interval = 100,
      fromEnd = true
      // Empezar desde el final del archivo
    } = options;
    let lastSize = 0;
    let lastPosition = 0;
    if (fs7.existsSync(filePath)) {
      const stats = fs7.statSync(filePath);
      lastSize = stats.size;
      lastPosition = fromEnd ? stats.size : 0;
    }
    const listener = async (curr, prev) => {
      try {
        if (curr.size > lastSize) {
          const stream = fs7.createReadStream(filePath, {
            start: lastPosition,
            end: curr.size - 1,
            encoding: "utf8"
          });
          let buffer = "";
          stream.on("data", (chunk) => {
            buffer += chunk;
            const lines = buffer.split("\n");
            for (let i = 0; i < lines.length - 1; i++) {
              if (lines[i].trim()) {
                callback(lines[i].trim(), filePath);
              }
            }
            buffer = lines[lines.length - 1];
          });
          stream.on("end", () => {
            if (buffer.trim()) {
              callback(buffer.trim(), filePath);
            }
          });
          lastSize = curr.size;
          lastPosition = curr.size;
        }
      } catch (error) {
        console.error("Error en watcher:", error);
      }
    };
    fs7.watchFile(filePath, { persistent, interval }, listener);
    const watcherId = `${filePath}_${Date.now()}`;
    _FileLogger.watchers.set(watcherId, listener);
    return {
      id: watcherId,
      stop: () => {
        const storedListener = _FileLogger.watchers.get(watcherId);
        if (storedListener) {
          fs7.unwatchFile(filePath, storedListener);
          _FileLogger.watchers.delete(watcherId);
        }
      },
      isWatching: () => _FileLogger.watchers.has(watcherId)
    };
  }
  /**
   * Detiene todos los watchers activos
   */
  static stopAllWatchers() {
    for (const [watcherId] of _FileLogger.watchers) {
      const filePath = watcherId.split("_")[0];
      fs7.unwatchFile(filePath);
    }
    _FileLogger.watchers.clear();
  }
  /**
   * Métodos de conveniencia para diferentes niveles de log
   */
  static async info(filePath, message) {
    return this.write(filePath, message, "INFO");
  }
  static async error(filePath, message) {
    return this.write(filePath, message, "ERROR");
  }
  static async warn(filePath, message) {
    return this.write(filePath, message, "WARN");
  }
  static async debug(filePath, message) {
    return this.write(filePath, message, "DEBUG");
  }
  /**
   * Limpia logs antiguos
   * @param filePath - Ruta del archivo de log
   * @param maxLines - Máximo número de líneas a mantener
   */
  static async cleanup(filePath, maxLines = 1e3) {
    try {
      const lines = await this.read(filePath, { asArray: true });
      if (lines.length > maxLines) {
        const keepLines = lines.slice(-maxLines);
        const content = keepLines.join("\n") + "\n";
        await fs7.promises.writeFile(filePath, content, "utf8");
        return lines.length - maxLines;
      }
      return 0;
    } catch (error) {
      console.error("Error limpiando logs:", error);
      throw error;
    }
  }
  /**
   * Elimina un archivo de log
   * @param filePath - Ruta del archivo de log a eliminar
   */
  static async deleteLogFile(filePath) {
    try {
      if (fs7.existsSync(filePath)) {
        await fs7.promises.unlink(filePath);
        return true;
      }
      return false;
    } catch (error) {
      console.error("Error eliminando archivo de log:", error);
      throw error;
    }
  }
  /**
   * Elimina múltiples archivos de log
   * @param filePaths - Array de rutas de archivos de log a eliminar
   */
  static async deleteLogFiles(filePaths) {
    const results = [];
    for (const filePath of filePaths) {
      try {
        const deleted = await this.deleteLogFile(filePath);
        results.push({ filePath, deleted, error: null });
      } catch (error) {
        results.push({ filePath, deleted: false, error: error.message });
      }
    }
    return results;
  }
};

// src/lib/Processors.ts
var ComputedFieldProcessor = class {
  static async getComputedFields(name) {
    let computedFields = [];
    if (await DbConfig_default.ifExist()) {
      await DbConfig_default.connect();
      try {
        const tableExistsQuery = `SELECT name FROM sqlite_master WHERE type='table' AND name='dbcube_computes_config'`;
        const tableExistsResult = await DbConfig_default.query(tableExistsQuery);
        if (tableExistsResult.status === "success" && tableExistsResult.data && tableExistsResult.data.length > 0) {
          const queryComputes = await DbConfig_default.queryWithParameters(`SELECT * FROM dbcube_computes_config WHERE database_ref=?`, [name]);
          computedFields = queryComputes.data;
        } else {
          computedFields = [];
        }
      } catch (error) {
        console.error("Error fetching computed fields:", error);
        computedFields = [];
      }
      await DbConfig_default.disconnect();
    }
    return computedFields;
  }
  /**
   * Processes computed field instruction and returns the computed value
   * @param instruction - The @compute instruction
   * @param rowData - The row data containing column values
   * @returns The computed value or null if there's an error
   */
  static processInstruction(instruction, rowData) {
    try {
      const functionMatch = instruction.match(/@compute\s*\(\s*\(\s*\)\s*=>\s*{([\s\S]*?)}\s*\)/);
      if (!functionMatch) {
        throw new Error("Invalid @compute instruction format");
      }
      let functionBody = functionMatch[1].trim();
      if (functionBody.endsWith(";")) {
        functionBody = functionBody.slice(0, -1).trim();
      }
      functionBody = functionBody.replace(/@column\(([^)]+)\)/g, (_match, columnName) => {
        const cleanColumnName = columnName.trim().replace(/['"]/g, "");
        return `__row[${JSON.stringify(cleanColumnName)}]`;
      });
      const computeFunction = new Function("__row", functionBody);
      return computeFunction(rowData);
    } catch (error) {
      console.error("Error processing computed field:", error);
      return null;
    }
  }
  /**
   * Extracts column dependencies from a computed field instruction
   * @param instruction - The @compute instruction
   * @returns Array of column names that this computed field depends on
   */
  static extractDependencies(instruction) {
    const dependencies = [];
    const columnMatches = instruction.match(/@column\(([^)]+)\)/g);
    if (columnMatches) {
      for (const match of columnMatches) {
        const innerMatch = match.match(/@column\(([^)]+)\)/);
        if (innerMatch) {
          const columnName = innerMatch[1].trim().replace(/['"]/g, "");
          if (!dependencies.includes(columnName)) {
            dependencies.push(columnName);
          }
        }
      }
    }
    return dependencies;
  }
  /**
   * Adds computed fields to an array of data objects based on configuration array
   * @param data - Array of data objects
   * @param computedConfigs - Array of computed field configurations
   * @returns Array with the new computed fields added
   */
  static computedFields(data, computedConfigs) {
    return data.map((item, index) => {
      let processedItem = { ...item };
      computedConfigs.forEach((config) => {
        try {
          const { column: fieldName, type: type2, instruction } = config;
          let processedExpression = instruction.replace(
            /@column\(([^)]+)\)/g,
            (match, columnName) => {
              const cleanColumnName = columnName.replace(/['"]/g, "");
              return `__row[${JSON.stringify(cleanColumnName)}]`;
            }
          );
          const computeMatch = processedExpression.match(/@compute\s*\(\s*\(\s*\)\s*=>\s*\{(.*)\}\s*\)/s);
          if (!computeMatch) {
            throw new Error(`Formato de @compute inv\xE1lido para campo ${fieldName}`);
          }
          const functionBody = computeMatch[1];
          const computeFunction = new Function("__row", functionBody);
          let result = computeFunction(processedItem);
          result = convertToType(result, type2);
          processedItem[fieldName] = result;
        } catch (error) {
          console.error(`Error procesando campo ${config.column} en item ${index}:`, error);
          processedItem[config.column] = null;
        }
      });
      return processedItem;
    });
  }
};
var TableProcessor = class {
  static async generateAlterQueries(nowQuery, dbType, tableName, database_ref) {
    function parseCreateTableQuery(query, dbType2) {
      const cleanQuery = query.trim().replace(/\s+/g, " ");
      const tableNameMatch = cleanQuery.match(/CREATE TABLE (?:IF NOT EXISTS )?(\w+)/i);
      if (!tableNameMatch) {
        throw new Error("No se pudo extraer el nombre de la tabla");
      }
      const tableName2 = tableNameMatch[1];
      const columnsMatch = cleanQuery.match(/\((.+)\)/);
      if (!columnsMatch) {
        throw new Error("No se pudieron extraer las definiciones de columnas");
      }
      const columnsString = columnsMatch[1];
      const columnDefinitions = columnsString.split(",").map((def) => def.trim());
      const columns = [];
      for (const def of columnDefinitions) {
        if (def.toUpperCase().startsWith("PRIMARY KEY") || def.toUpperCase().startsWith("FOREIGN KEY") || def.toUpperCase().startsWith("CONSTRAINT")) {
          continue;
        }
        const parts = def.split(/\s+/);
        if (parts.length < 2) continue;
        const columnName = parts[0];
        const columnType = parts[1];
        const column = {
          name: columnName,
          type: columnType,
          nullable: !def.toUpperCase().includes("NOT NULL"),
          primaryKey: def.toUpperCase().includes("PRIMARY KEY"),
          autoIncrement: def.toUpperCase().includes("AUTOINCREMENT") || def.toUpperCase().includes("AUTO_INCREMENT")
        };
        columns.push(column);
      }
      return { tableName: tableName2, columns };
    }
    function generateMySQLQueries(nowColumns, oldColumns, tableName2) {
      const queries = [];
      for (const [columnName] of oldColumns) {
        if (!nowColumns.has(columnName)) {
          queries.push(`ALTER TABLE ${tableName2} DROP COLUMN ${columnName};`);
        }
      }
      for (const [columnName, column] of nowColumns) {
        const oldColumn = oldColumns.get(columnName);
        if (!oldColumn) {
          const columnDef = buildColumnDefinition(column, "mysql");
          queries.push(`ALTER TABLE ${tableName2} ADD COLUMN ${columnDef};`);
        } else if (!columnsEqual(column, oldColumn)) {
          const columnDef = buildColumnDefinition(column, "mysql");
          queries.push(`ALTER TABLE ${tableName2} MODIFY COLUMN ${columnDef};`);
        }
      }
      return queries;
    }
    function generatePostgreSQLQueries(nowColumns, oldColumns, tableName2) {
      const queries = [];
      for (const [columnName] of oldColumns) {
        if (!nowColumns.has(columnName)) {
          queries.push(`ALTER TABLE ${tableName2} DROP COLUMN ${columnName};`);
        }
      }
      for (const [columnName, column] of nowColumns) {
        const oldColumn = oldColumns.get(columnName);
        if (!oldColumn) {
          const columnDef = buildColumnDefinition(column, "postgres");
          queries.push(`ALTER TABLE ${tableName2} ADD COLUMN ${columnDef};`);
        } else if (!columnsEqual(column, oldColumn)) {
          if (column.type !== oldColumn.type) {
            const pgType = column.type.replace("INTEGER", "INT").replace("TEXT", "VARCHAR");
            queries.push(`ALTER TABLE ${tableName2} ALTER COLUMN ${columnName} TYPE ${pgType};`);
          }
          if (column.nullable !== oldColumn.nullable) {
            const constraint = column.nullable ? "DROP NOT NULL" : "SET NOT NULL";
            queries.push(`ALTER TABLE ${tableName2} ALTER COLUMN ${columnName} ${constraint};`);
          }
        }
      }
      return queries;
    }
    function generateSQLiteQueries(nowSchema, oldSchema, tableName2) {
      const queries = [];
      if (schemasEqual(nowSchema, oldSchema)) {
        return queries;
      }
      const tempTableName = `${tableName2}_temp_${Date.now()}`;
      const createTempQuery = buildCreateTableQuery(nowSchema, "sqlite", tempTableName);
      queries.push(createTempQuery);
      const commonColumns = nowSchema.columns.filter((col) => oldSchema.columns.some((oldCol) => oldCol.name.toLowerCase() === col.name.toLowerCase())).map((col) => col.name);
      if (commonColumns.length > 0) {
        const columnsList = commonColumns.join(", ");
        queries.push(`INSERT INTO ${tempTableName} (${columnsList}) SELECT ${columnsList} FROM ${tableName2};`);
      }
      queries.push(`DROP TABLE ${tableName2};`);
      queries.push(`ALTER TABLE ${tempTableName} RENAME TO ${tableName2};`);
      return queries;
    }
    function generateMongoDBQueries(nowColumns, oldColumns, collectionName) {
      const queries = [];
      const fieldsToRemove = [];
      for (const [fieldName] of oldColumns) {
        if (!nowColumns.has(fieldName)) {
          fieldsToRemove.push(fieldName);
        }
      }
      if (fieldsToRemove.length > 0) {
        const unsetFields = fieldsToRemove.reduce((acc, field) => {
          acc[field] = "";
          return acc;
        }, {});
        queries.push(`db.${collectionName}.updateMany({}, { $unset: ${JSON.stringify(unsetFields)} });`);
      }
      const fieldsToAdd = {};
      for (const [fieldName, column] of nowColumns) {
        if (!oldColumns.has(fieldName)) {
          let defaultValue = null;
          if (!column.nullable) {
            switch (column.type.toUpperCase()) {
              case "INTEGER":
                defaultValue = 0;
                break;
              case "TEXT":
              case "VARCHAR":
                defaultValue = "";
                break;
              default:
                defaultValue = null;
            }
          }
          fieldsToAdd[fieldName] = defaultValue;
        }
      }
      if (Object.keys(fieldsToAdd).length > 0) {
        queries.push(`db.${collectionName}.updateMany({}, { $set: ${JSON.stringify(fieldsToAdd)} });`);
      }
      return queries;
    }
    function buildColumnDefinition(column, dbType2) {
      let def = `${column.name} ${column.type}`;
      if (column.primaryKey) {
        def += " PRIMARY KEY";
      }
      if (column.autoIncrement) {
        if (dbType2 === "mysql") {
          def += " AUTO_INCREMENT";
        } else if (dbType2 === "sqlite") {
          def += " AUTOINCREMENT";
        } else if (dbType2 === "postgres") {
          def = def.replace(column.type, "SERIAL");
        }
      }
      if (!column.nullable) {
        def += " NOT NULL";
      }
      return def;
    }
    function buildCreateTableQuery(schema, dbType2, tableName2) {
      const name = tableName2 || schema.tableName;
      const columnDefs = schema.columns.map((col) => buildColumnDefinition(col, dbType2));
      return `CREATE TABLE ${name} (${columnDefs.join(", ")});`;
    }
    function columnsEqual(col1, col2) {
      return col1.name.toLowerCase() === col2.name.toLowerCase() && col1.type === col2.type && col1.nullable === col2.nullable && col1.primaryKey === col2.primaryKey && col1.autoIncrement === col2.autoIncrement;
    }
    function schemasEqual(schema1, schema2) {
      if (schema1.columns.length !== schema2.columns.length) {
        return false;
      }
      const cols1 = new Map(schema1.columns.map((col) => [col.name.toLowerCase(), col]));
      const cols2 = new Map(schema2.columns.map((col) => [col.name.toLowerCase(), col]));
      for (const [name, col1] of cols1) {
        const col2 = cols2.get(name);
        if (!col2 || !columnsEqual(col1, col2)) {
          return false;
        }
      }
      return true;
    }
    if (await DbConfig_default.ifExist()) {
      await DbConfig_default.connect();
      try {
        const tableExistsQuery = `SELECT name FROM sqlite_master WHERE type='table' AND name='dbcube_schemas_config'`;
        const tableExistsResult = await DbConfig_default.query(tableExistsQuery);
        if (tableExistsResult.status === "success" && tableExistsResult.data && tableExistsResult.data.length > 0) {
          const queryComputes = await DbConfig_default.queryWithParameters(`SELECT * FROM dbcube_schemas_config WHERE table_ref=? AND database_ref=?`, [tableName, database_ref]);
          const oldQuery = queryComputes.data[0];
          if (!oldQuery || !oldQuery.struct) {
            await DbConfig_default.disconnect();
            return [nowQuery];
          }
          const nowSchema = parseCreateTableQuery(nowQuery, dbType);
          const oldSchema = parseCreateTableQuery(oldQuery.struct, dbType);
          if (nowSchema.tableName.toLowerCase() !== tableName.toLowerCase()) {
            throw new Error(`El nombre de tabla en la query (${nowSchema.tableName}) no coincide con el par\xE1metro (${tableName})`);
          }
          const nowColumns = new Map(nowSchema.columns.map((col) => [col.name.toLowerCase(), col]));
          const oldColumns = new Map(oldSchema.columns.map((col) => [col.name.toLowerCase(), col]));
          switch (dbType) {
            case "mysql":
              return generateMySQLQueries(nowColumns, oldColumns, tableName);
            case "postgres":
              return generatePostgreSQLQueries(nowColumns, oldColumns, tableName);
            case "sqlite":
              return generateSQLiteQueries(nowSchema, oldSchema, tableName);
            case "mongodb":
              return generateMongoDBQueries(nowColumns, oldColumns, tableName);
            default:
              throw new Error(`Tipo de base de datos no soportado: ${dbType}`);
          }
        }
      } catch (error) {
        console.error("Error fetching computed fields:", error);
      }
      await DbConfig_default.disconnect();
      return [nowQuery];
    } else {
      return [nowQuery];
    }
  }
  static async saveQuery(table_ref, database_ref, struct) {
    if (await DbConfig_default.ifExist()) {
      await DbConfig_default.connect();
      try {
        await DbConfig_default.queryWithParameters(`DELETE FROM dbcube_schemas_config WHERE table_ref=? AND database_ref=?`, [table_ref, database_ref]);
        const tableExistsQuery = `INSERT INTO dbcube_schemas_config (table_ref, database_ref, struct) VALUES (?, ?, ?)`;
        await DbConfig_default.queryWithParameters(tableExistsQuery, [table_ref, database_ref, struct]);
      } catch (error) {
        console.error("Error fetching computed fields:", error);
      }
      await DbConfig_default.disconnect();
    }
  }
};
var TriggerProcessor = class {
  static async getTriggers(name) {
    let triggers = [];
    if (await DbConfig_default.ifExist()) {
      await DbConfig_default.connect();
      try {
        const tableExistsQuery = `SELECT name FROM sqlite_master WHERE type='table' AND name='dbcube_triggers_config'`;
        const tableExistsResult = await DbConfig_default.query(tableExistsQuery);
        if (tableExistsResult.status === "success" && tableExistsResult.data && tableExistsResult.data.length > 0) {
          const queryComputes = await DbConfig_default.queryWithParameters(`SELECT * FROM dbcube_triggers_config WHERE database_ref=?`, [name]);
          triggers = queryComputes.data;
        } else {
          triggers = [];
        }
      } catch (error) {
        console.error("Error fetching computed fields:", error);
        triggers = [];
      }
      await DbConfig_default.disconnect();
    }
    return triggers;
  }
};
function convertToType(value, type2) {
  switch (type2) {
    case "string":
      return String(value);
    case "number":
      const num = Number(value);
      return isNaN(num) ? 0 : num;
    case "boolean":
      return Boolean(value);
    case "date":
      return value instanceof Date ? value : new Date(value);
    case "object":
      return value;
    default:
      return value;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Arquitecture,
  Binary,
  ComputedFieldProcessor,
  Config,
  DaemonClient,
  DbConfig,
  Engine,
  FileLogger,
  QueryEngine,
  TableProcessor,
  TriggerProcessor
});
