// @dbcube/cli/src/commands/migrate/lib.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

/**
 * Migration history for .alter.cube files.
 * Stored in dbcube/migrations.json so the history travels with the repo.
 */
const HISTORY_PATH = () => path.join(process.cwd(), 'dbcube', 'migrations.json');

function loadHistory() {
    try {
        return JSON.parse(fs.readFileSync(HISTORY_PATH(), 'utf8'));
    } catch {
        return { applied: [], lastBatch: 0 };
    }
}

function saveHistory(history) {
    fs.mkdirSync(path.dirname(HISTORY_PATH()), { recursive: true });
    fs.writeFileSync(HISTORY_PATH(), JSON.stringify(history, null, 2), 'utf8');
}

function hashFile(filePath) {
    return crypto.createHash('sha1').update(fs.readFileSync(filePath)).digest('hex');
}

function collectAlterFiles(dir = path.join(process.cwd(), 'dbcube'), acc = []) {
    if (!fs.existsSync(dir)) return acc;
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) {
            if (!['triggers', 'logs', 'node_modules', 'applied'].includes(entry.name)) collectAlterFiles(full, acc);
        } else if (entry.name.endsWith('.alter.cube')) {
            acc.push(full);
        }
    }
    return acc;
}

/** Splits an .alter.cube into header (@database/@table) + directive blocks. */
function parseAlterFile(content) {
    const header = [];
    const dbMatch = content.match(/@database\s*\([^)]*\)\s*;/);
    const tableMatch = content.match(/@table\s*\([^)]*\)\s*;/);
    if (dbMatch) header.push(dbMatch[0]);
    if (tableMatch) header.push(tableMatch[0]);

    const directives = [];
    const regex = /@(changeName|addColumn|deleteColumn|renameColumn|changeType|changeLength|changeDefault|changeOptions|changeEnumValues)\s*\(([\s\S]*?)\)\s*;/g;
    let m;
    while ((m = regex.exec(content)) !== null) {
        directives.push({ name: m[1], arg: m[2].trim(), raw: m[0] });
    }
    return { header, directives };
}

function swapLastNew(arg) {
    // Swaps every {last: "x"; new: "y"} pair inside the directive argument
    return arg.replace(
        /last\s*:\s*"([^"]*)"\s*;\s*new\s*:\s*"([^"]*)"/g,
        (_, last, next) => `last: "${next}"; new: "${last}"`
    );
}

/** Inverts one directive. Returns a string (reverse cube code) or null if irreversible. */
function invertDirective(d) {
    switch (d.name) {
        case 'changeName':
        case 'renameColumn':
        case 'changeType':
        case 'changeLength':
        case 'changeDefault':
            return `@${d.name}(${swapLastNew(d.arg)});`;
        case 'addColumn': {
            // Reverse of adding columns = deleting them
            const cols = [...d.arg.matchAll(/^\s*(\w+)\s*:\s*\{/gm)].map(m => m[1]);
            // Top-level columns only: the first identifier after { or after a top-level };
            const topCols = [...d.arg.matchAll(/(?:^|\};)\s*(\w+)\s*:\s*\{/g)].map(m => m[1]);
            const names = topCols.length > 0 ? topCols : cols;
            if (names.length === 0) return null;
            return names.map(c => `@deleteColumn("${c}");`).join('\n');
        }
        case 'changeOptions': {
            // Swap add/remove arrays
            const swapped = d.arg
                .replace(/\badd\s*:/g, '__TMP__:')
                .replace(/\bremove\s*:/g, 'add:')
                .replace(/__TMP__:/g, 'remove:');
            return `@changeOptions(${swapped});`;
        }
        case 'deleteColumn':
        case 'changeEnumValues':
            return null; // irreversible: dropped data / unknown previous values
        default:
            return null;
    }
}

/**
 * Builds the full reverse .alter.cube content for a file.
 * Returns { reverse: string | null, irreversible: string[] }.
 */
function buildReverse(content) {
    const { header, directives } = parseAlterFile(content);
    const irreversible = [];
    const reversed = [];

    for (const d of [...directives].reverse()) {
        const inv = invertDirective(d);
        if (inv === null) {
            irreversible.push(`@${d.name}`);
        } else {
            reversed.push(inv);
        }
    }

    if (irreversible.length > 0) return { reverse: null, irreversible };
    return {
        reverse: header.join('\n') + '\n\n' + reversed.join('\n\n') + '\n',
        irreversible: [],
    };
}

module.exports = { loadHistory, saveHistory, hashFile, collectAlterFiles, buildReverse };
