// @dbcube/cli/src/commands/generate.js
const fs = require('fs');
const path = require('path');
const chalk = require('chalk');

/**
 * dbcube generate — Generates TypeScript interfaces from .table.cube files.
 * Output: dbcube/types.ts (one interface per table + DbSchema map).
 *
 * Usage with the query builder:
 *   import type { User } from './dbcube/types';
 *   const users = await db.table('users').get() as User[];
 */
const TYPE_MAP = {
    int: 'number', integer: 'number', bigint: 'number', tinyint: 'number',
    float: 'number', double: 'number', decimal: 'number',
    varchar: 'string', string: 'string', text: 'string', uuid: 'string',
    date: 'string', datetime: 'string', timestamp: 'string',
    boolean: 'boolean', bool: 'boolean',
    json: 'any',
};

function collectTableCubes(dir, acc = []) {
    if (!fs.existsSync(dir)) return acc;
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) {
            if (!['triggers', 'logs', 'node_modules'].includes(entry.name)) collectTableCubes(full, acc);
        } else if (entry.name.endsWith('.table.cube')) {
            acc.push(full);
        }
    }
    return acc;
}

function pascalCase(name) {
    return name.split(/[_\-\s]+/).map(w => w.charAt(0).toUpperCase() + w.slice(1)).join('');
}

function singularize(word) {
    if (word.endsWith('ies')) return word.slice(0, -3) + 'y';
    if (word.endsWith('ses')) return word.slice(0, -2);
    if (word.endsWith('s') && !word.endsWith('ss')) return word.slice(0, -1);
    return word;
}

/** Parses the @columns block of a .table.cube into [{name, tsType, optional, comment}] */
function parseCube(content) {
    const nameMatch = content.match(/@meta\s*\(\s*\{[\s\S]*?name\s*:\s*"([^"]+)"/);
    const dbMatch = content.match(/@database\s*\(\s*"([^"]+)"\s*\)/);
    const columnsMatch = content.match(/@columns\s*\(\s*\{([\s\S]*)\}\s*\)\s*;?\s*$/m)
        || content.match(/@columns\s*\(\s*\{([\s\S]*?)\}\s*\)\s*;/);
    if (!columnsMatch) return null;

    const body = columnsMatch[1];
    const columns = [];
    // Match each top-level "col: { ... };" block (tolerates one nesting level: foreign/@compute braces)
    const colRegex = /(\w+)\s*:\s*\{((?:[^{}]|\{(?:[^{}]|\{[^{}]*\})*\})*?)\}\s*;/g;
    let m;
    while ((m = colRegex.exec(body)) !== null) {
        const colName = m[1];
        const def = m[2];
        const typeMatch = def.match(/type\s*:\s*"([^"]+)"/);
        if (!typeMatch) continue;
        const cubeType = typeMatch[1].toLowerCase();

        let tsType = TYPE_MAP[cubeType] ?? 'any';
        if (cubeType === 'enum') {
            const enumMatch = def.match(/enumValues\s*:\s*\[([^\]]*)\]/);
            if (enumMatch) {
                const values = enumMatch[1].match(/"([^"]*)"/g);
                if (values && values.length > 0) tsType = values.join(' | ');
            }
        }

        const options = (def.match(/options\s*:\s*\[([^\]]*)\]/) || [])[1] || '';
        const notNull = /"not null"/.test(options) || /"primary"/.test(options);
        const hasDefault = /defaultValue\s*:/.test(def) || /"autoincrement"/.test(options);
        const isComputed = /@compute/.test(def);

        columns.push({
            name: colName,
            tsType: notNull ? tsType : `${tsType} | null`,
            optional: !notNull || hasDefault || isComputed,
            comment: isComputed ? 'computed field (read-only, requires useComputes())' : null,
        });
    }

    return {
        table: nameMatch ? nameMatch[1] : null,
        database: dbMatch ? dbMatch[1] : null,
        columns,
    };
}

async function main() {
    const cubesDir = path.join(process.cwd(), 'dbcube');
    const files = collectTableCubes(cubesDir);

    if (files.length === 0) {
        console.log(`${chalk.yellow('⚠')} No .table.cube files found in dbcube/`);
        process.exit(0);
    }

    console.log(`\n🧬 ${chalk.green('Generating TypeScript types from')} ${files.length} cube(s)...\n`);

    let out = `// ⚠ AUTO-GENERATED by \`dbcube generate\` — do not edit by hand.\n`;
    out += `// Source of truth: your .table.cube files. Re-run after schema changes.\n\n`;

    const schemaEntries = [];

    for (const file of files) {
        const content = fs.readFileSync(file, 'utf8');
        const parsed = parseCube(content);
        const tableName = (parsed && parsed.table) || path.basename(file, '.table.cube');

        if (!parsed || parsed.columns.length === 0) {
            console.log(`  ${chalk.yellow('⚠')} ${path.relative(process.cwd(), file)} — could not parse columns, skipped`);
            continue;
        }

        const ifaceName = pascalCase(singularize(tableName));
        out += `/** Table \`${tableName}\`${parsed.database ? ` · database \`${parsed.database}\`` : ''} */\n`;
        out += `export interface ${ifaceName} {\n`;
        for (const col of parsed.columns) {
            if (col.comment) out += `    /** ${col.comment} */\n`;
            out += `    ${col.name}: ${col.tsType};\n`;
        }
        out += `}\n\n`;

        out += `/** Shape accepted by insert()/update() for \`${tableName}\` */\n`;
        out += `export interface New${ifaceName} {\n`;
        for (const col of parsed.columns) {
            if (col.comment) continue; // computed fields are not writable
            out += `    ${col.name}${col.optional ? '?' : ''}: ${col.tsType};\n`;
        }
        out += `}\n\n`;

        schemaEntries.push({ tableName, ifaceName });
        console.log(`  ${chalk.green('✓')} ${tableName} → ${ifaceName}, New${ifaceName}`);
    }

    out += `/** Map of table name → row type. */\n`;
    out += `export interface DbSchema {\n`;
    for (const { tableName, ifaceName } of schemaEntries) {
        out += `    ${tableName}: ${ifaceName};\n`;
    }
    out += `}\n`;

    const outPath = path.join(cubesDir, 'types.ts');
    fs.writeFileSync(outPath, out, 'utf8');
    console.log(`\n${chalk.green('✓')} Types written to ${chalk.bold('dbcube/types.ts')}\n`);
    console.log(`${chalk.cyan('Usage:')}`);
    console.log(`  import type { ${schemaEntries[0] ? schemaEntries[0].ifaceName : 'User'} } from './dbcube/types';`);
    console.log(`  const rows = await db.table('${schemaEntries[0] ? schemaEntries[0].tableName : 'users'}').get() as ${schemaEntries[0] ? schemaEntries[0].ifaceName : 'User'}[];\n`);
}

main().catch(error => {
    console.error('Error fatal:', error);
    process.exit(1);
});
