// @dbcube/cli/src/utils/ConfigFileUtils.js
const fs = require('fs');
const path = require('path');
const EnvLoader = require('./EnvLoader');

/**
 * Detecta si el proyecto está marcado como ESM ("type": "module" en
 * package.json), lo que hace que dbcube.config.js (CommonJS) no se pueda
 * cargar con require(). Devuelve true si es ESM.
 */
function projectIsESM() {
    try {
        const pkgPath = path.resolve(process.cwd(), 'package.json');
        if (!fs.existsSync(pkgPath)) return false;
        const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
        return pkg.type === 'module';
    } catch {
        return false;
    }
}

/**
 * Clase para manejar operaciones sobre el archivo de configuración
 */
class ConfigFileUtils {
    /**
     * Obtiene la ruta al archivo de configuración
     * @returns {string} - Ruta al archivo de configuración
     */
    static getConfigPath() {
        const rootPath = path.resolve(process.cwd());
        const cjsPath = path.join(rootPath, 'dbcube.config.cjs');
        const jsPath = path.join(rootPath, 'dbcube.config.js');
        // Si ya existe alguno, ese manda (preferimos .cjs).
        if (fs.existsSync(cjsPath)) return cjsPath;
        if (fs.existsSync(jsPath)) return jsPath;
        // Si no existe ninguno (se va a crear): en proyectos ESM
        // ("type": "module") debe ser .cjs para poder cargarse con require().
        return projectIsESM() ? cjsPath : jsPath;
    }

    /**
     * Obtiene los nombres de las bases de datos configuradas en dbcube.config.js
     * @returns {Promise<Array<string>>} - Array con los nombres de las bases de datos
     */
    static async getConfiguredDatabases() {
        // Sin try/catch que oculte el error: loadDatabasesFromPath ya lanza
        // mensajes explícitos y accionables. (Antes, `new Error('msg', error)`
        // descartaba el segundo argumento y se perdía la causa real.)
        const configPath = this.getConfigPath();

        if (!fs.existsSync(configPath)) {
            ConfigFileUtils.crearArchivoConfig(configPath);
        }

        return this.loadDatabasesFromPath(configPath);
    }
    
    /**
     * Carga las bases de datos desde una ruta específica
     * @param {string} configPath - Ruta al archivo de configuración
     * @returns {Array<string>} - Nombres de las bases de datos
     */
    static loadDatabasesFromPath(configPath) {
        // Carga automática de .env (cargador propio de DBCube): el usuario no
        // necesita instalar dotenv ni escribir require("dotenv") en su config.
        EnvLoader.load();

        let config;
        try {
            // Borrar la caché del módulo para asegurarnos de cargar la versión más reciente
            delete require.cache[require.resolve(configPath)];
            // Cargar el archivo de configuración
            config = require(configPath);
        } catch (error) {
            const missing = (error.message.match(/Cannot find module '([^']+)'/) || [])[1];
            // ¿El módulo que no se encuentra es el PROPIO config? Entonces el
            // problema es de carga (típicamente ESM). Si es OTRO módulo, es una
            // dependencia del config que falta (p. ej. dotenv).
            const missingIsConfig = missing && (configPath.includes(missing) || missing.includes('dbcube.config'));

            // Caso 1: falta una dependencia que el config requiere.
            if (error.code === 'MODULE_NOT_FOUND' && missing && !missingIsConfig) {
                throw new Error(
                    `dbcube.config.js requiere un módulo que no está instalado: "${missing}".\n` +
                    `   Instálalo (p. ej. npm install ${missing}).\n` +
                    `   Nota: DBCube ya carga el .env automáticamente — puedes quitar require("dotenv") de tu config.`
                );
            }
            // Caso 2: proyecto ESM ("type": "module") → require() del config falla
            // o el .js se interpreta como módulo ES y `module.exports` no aplica.
            if (projectIsESM() || error.code === 'ERR_REQUIRE_ESM') {
                throw new Error(
                    `No se pudo cargar dbcube.config.js porque el proyecto está marcado como ESM ` +
                    `("type": "module" en package.json), pero la configuración usa CommonJS (module.exports).\n` +
                    `   Solución: renombra el archivo a "dbcube.config.cjs", o quita "type": "module" de package.json.`
                );
            }
            // Cualquier otro error de carga: mostrarlo tal cual.
            throw new Error(`No se pudo cargar la configuración desde ${configPath}:\n   ${error.message}`);
        }

        // Validar que la config exporte una función (firma esperada).
        if (typeof config !== 'function') {
            const hint = projectIsESM()
                ? ` El proyecto es ESM ("type": "module"): renombra el archivo a "dbcube.config.cjs" o quita "type": "module".`
                : ` Debe exportar una función: module.exports = function (config) { config.set({ databases: {...} }); }.`;
            throw new Error(`dbcube.config.js no exporta una configuración válida.${hint}`);
        }

        // Crear una instancia temporal de Config para obtener los datos
        const tempConfig = {
            databases: {},
            data: {},
            set: function(configData) {
                this.data = configData;
                if (configData.databases) {
                    this.databases = configData.databases;
                }
            },
            getAllDatabases: function() {
                return this.databases;
            }
        };

        // Ejecutar la función de configuración
        try {
            config(tempConfig);
        } catch (error) {
            throw new Error(`dbcube.config.js lanzó un error al evaluarse:\n   ${error.message}`);
        }

        const allDatabases = tempConfig.getAllDatabases();
        if (!allDatabases || Object.keys(allDatabases).length === 0) {
            throw new Error(
                `dbcube.config.js no define ninguna base de datos.\n` +
                `   Añade al menos una en config.set({ databases: { miBase: { type, config } } }).`
            );
        }

        const databaseNames = Object.keys(allDatabases);
        const arrayDatabases = [];
        for (const database of databaseNames) {
            arrayDatabases.push({
                name: database,
                type: allDatabases[database].type
            });
        }
        return arrayDatabases;
    }

    /**
     * Agrega una nueva configuración de base de datos al archivo de configuración
     * @param {string} databaseName - Nombre de la base de datos
     * @param {string} databaseType - Tipo de la base de datos (mysql, postgresql, etc.)
     * @param {Object} databaseConfig - Configuración de la base de datos
     * @returns {Promise<boolean>} - Retorna true si se agregó correctamente
     */
    static async addDatabaseConfig(databaseName, databaseType, databaseConfig) {
        try {
            // Ruta al archivo de configuración
            const configPath = this.getConfigPath();
            
            // Verificar si el archivo existe
            if (!fs.existsSync(configPath)) {
                ConfigFileUtils.crearArchivoConfig(configPath);
            }
            
            // Leer el archivo de configuración
            let configContent = fs.readFileSync(configPath, 'utf8');
            
            // Convertir los valores de PORT a número si es necesario
            if (databaseConfig.PORT && typeof databaseConfig.PORT === 'string') {
                databaseConfig.PORT = parseInt(databaseConfig.PORT, 10);
            }
            
            // Crear el texto de la nueva configuración
            const newDatabaseConfigText = this.createDatabaseConfigText(databaseName, databaseType, databaseConfig);
            
            // Buscar la posición donde insertar la nueva configuración
            const insertPosition = configContent.indexOf('databases: {') + 'databases: {'.length;
            if (insertPosition === -1 + 'databases: {'.length) {
                throw new Error('No se encontró la sección de bases de datos en el archivo de configuración');
            }
            
            // Insertar la nueva configuración
            configContent = configContent.slice(0, insertPosition) + 
                            newDatabaseConfigText + 
                            configContent.slice(insertPosition);
            
            // Escribir el archivo actualizado
            fs.writeFileSync(configPath, configContent, 'utf8');
            
            return true;
        } catch (error) {
            console.error('\nError al agregar la configuración de la base de datos:', error);
            return false;
        }
    }
    
    /**
     * Crea el texto de configuración para una base de datos
     * @param {string} databaseName - Nombre de la base de datos
     * @param {string} databaseType - Tipo de la base de datos
     * @param {Object} databaseConfig - Configuración de la base de datos
     * @returns {string} - Texto de configuración formateado
     */
    static createDatabaseConfigText(databaseName, databaseType, databaseConfig) {
        if(databaseType=="sqlite"){
            const arrayDataEnv = [
                "DBCUBE_"+databaseName.toUpperCase()+"_DATABASE"+"="+databaseConfig.DATABASE
            ];
            ConfigFileUtils.updateDotenv(arrayDataEnv)
            return `
                ${databaseName}: {
                    type: "${databaseType}",
                    config:{
                        DATABASE: process.env.${arrayDataEnv[0].split("=")[0]}
                    }
                },`;
        }else{
            const arrayDataEnv = [
                "DBCUBE_"+databaseName.toUpperCase()+"_HOST"+"="+databaseConfig.HOST,
                "DBCUBE_"+databaseName.toUpperCase()+"_PORT"+"="+databaseConfig.PORT,
                "DBCUBE_"+databaseName.toUpperCase()+"_DATABASE"+"="+databaseConfig.DATABASE,
                "DBCUBE_"+databaseName.toUpperCase()+"_USER"+"="+databaseConfig.USER,
                "DBCUBE_"+databaseName.toUpperCase()+"_PASSWORD"+"="+databaseConfig.PASSWORD,
            ];
            ConfigFileUtils.updateDotenv(arrayDataEnv)
            return `
                ${databaseName}: {
                    type: "${databaseType}",
                    config:{
                        HOST: process.env.${arrayDataEnv[0].split("=")[0]},
                        USER: process.env.${arrayDataEnv[3].split("=")[0]},
                        PASSWORD: process.env.${arrayDataEnv[4].split("=")[0]},
                        DATABASE: process.env.${arrayDataEnv[2].split("=")[0]},
                        PORT: parseInt(process.env.${arrayDataEnv[1].split("=")[0]}, 10)
                    }
                },`;
        }
    }

    /**
     * Crea o actualiza un archivo .env con un array de strings tipo "KEY=VALUE"
     * @param {string} ruta - Ruta donde se creará o modificará el archivo `.env`
     * @param {string[]} entries - Array de strings tipo "KEY=VALUE"
     */
    static updateDotenv(entries) {
        const rutaAbsoluta = path.resolve(process.cwd(), '.env');
        let contenidoExistente = '';
    
        try {
            // Verificar si la ruta existe y qué tipo de archivo es
            if (fs.existsSync(rutaAbsoluta)) {
                const stats = fs.statSync(rutaAbsoluta);
                
                if (stats.isDirectory()) {
                    // Si .env es un directorio, elimínalo y crea el archivo
                    console.warn(`⚠️  Se encontró un directorio llamado .env, eliminándolo para crear el archivo`);
                    fs.rmSync(rutaAbsoluta, { recursive: true, force: true });
                } else if (stats.isFile()) {
                    // Si es un archivo, leer su contenido
                    contenidoExistente = fs.readFileSync(rutaAbsoluta, 'utf8');
                }
            }
        } catch (error) {
            if (error.code === 'EISDIR') {
                console.warn(`⚠️  Conflicto detectado: .env existe como directorio, eliminándolo`);
                try {
                    fs.rmSync(rutaAbsoluta, { recursive: true, force: true });
                } catch (removeError) {
                    console.error(`❌ Error al eliminar directorio .env:`, removeError.message);
                    throw removeError;
                }
            } else {
                console.error(`❌ Error al leer .env:`, error.message);
                throw error;
            }
        }
    
        // Procesar contenido existente
        const lineas = contenidoExistente ? contenidoExistente.split('\n') : [];
        const mapaEnv = {};
    
        // Construir mapa de claves actuales
        for (let linea of lineas) {
            const match = linea.match(/^([\w.-]+)=(.*)$/);
            if (match) {
                const [, key, value] = match;
                mapaEnv[key] = value;
            }
        }
    
        // Procesar nuevas entradas
        for (const entry of entries) {
            const [key, ...valueParts] = entry.split('=');
            const value = valueParts.join('='); // En caso de que el valor contenga '='
            
            if (!key || value === undefined) {
                console.warn(`⚠️  Entrada inválida ignorada: ${entry}`);
                continue;
            }
    
            if (!mapaEnv.hasOwnProperty(key)) {
                // Agregar nueva línea si no existe
                lineas.push(`${key}=${value}`);
                console.log(`➕ Agregando nueva variable: ${key}`);
            } else if (mapaEnv[key] !== value) {
                // Actualizar línea existente
                const index = lineas.findIndex(line => line.startsWith(`${key}=`));
                if (index !== -1) {
                    lineas[index] = `${key}=${value}`;
                    console.log(`🔄 Actualizando variable: ${key}`);
                }
            }
            // Si existe y tiene el mismo valor, no se hace nada
        }
    
        try {
            // Filtrar líneas vacías al final y asegurar que termine con salto de línea
            const contenidoFinal = lineas.filter(linea => linea.trim() !== '').join('\n');
            
            // Escribir de nuevo el archivo .env
            fs.writeFileSync(rutaAbsoluta, contenidoFinal + '\n', 'utf8');
            console.log(`✅ .env actualizado en: ${rutaAbsoluta}`);
            
        } catch (error) {
            console.error(`❌ Error al escribir .env:`, error.message);
            throw error;
        }
    }

    /**
     * Crea un archivo de configuración base para dbcube en la ruta especificada.
     * @param {string} rutaDestino - Ruta completa donde se debe crear el archivo.
     */
    static crearArchivoConfig(rutaDestino) {
        // DBCube carga el .env automáticamente, así que el template ya NO
        // necesita require('dotenv') (que además requeriría instalar el paquete).
        const contenido = `// DBCube carga .env automáticamente — usa process.env.MI_VAR sin dotenv.
module.exports = function (config) {
    config.set({
        databases: {
        }
    });
};
`;

        // Asegurarse de que el directorio existe
        const dir = path.dirname(rutaDestino);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }

        // Escribir archivo
        fs.writeFileSync(rutaDestino, contenido, 'utf8');
        //console.log(`\n✅ Archivo creado en: ${rutaDestino}`);
    }
}

module.exports = ConfigFileUtils;
