// @dbcube/cli/src/utils/FileUtils.js
const fs = require('fs');
const path = require('path');

class FileUtils {
  /**
   * Verifica si un archivo existe (asincrónico).
   * @param {string} filePath - Ruta del archivo.
   * @returns {Promise<boolean>} - True si el archivo existe, false si no.
   */
  static async fileExists(filePath) {
    return new Promise((resolve) => {
      fs.access(path.resolve(filePath), fs.constants.F_OK, (err) => {
        resolve(!err);
      });
    });
  }

  /**
   * Verifica si un archivo existe (sincrónico).
   * @param {string} filePath - Ruta del archivo.
   * @returns {boolean} - True si el archivo existe, false si no.
   */
  static fileExistsSync(filePath) {
    try {
      fs.accessSync(path.resolve(filePath), fs.constants.F_OK);
      return true;
    } catch {
      return false;
    }
  }

  static extractDatabaseName(input) {
    const match = input.match(/@database\(["']?([\w-]+)["']?\)/);
    return match ? match[1] : null;
  }

   /**
   * Lee recursivamente archivos que terminan en un sufijo dado y los ordena numéricamente.
   * @param {string} dir - Directorio base (relativo o absoluto).
   * @param {string} suffix - Sufijo de archivo (como 'table.cube').
   * @returns {string[]} - Rutas absolutas de los archivos encontrados y ordenados.
   */
   static getCubeFilesRecursively(dir, suffix) {
    const baseDir = path.resolve(dir); // ✅ Asegura que sea absoluto
    const cubeFiles = [];

    function recurse(currentDir) {
      const entries = fs.readdirSync(currentDir, { withFileTypes: true });

      for (const entry of entries) {
        const fullPath = path.join(currentDir, entry.name);
        if (entry.isDirectory()) {
          recurse(fullPath);
        } else if (entry.isFile() && (entry.name.endsWith(suffix) || entry.name.includes(suffix))) {
          cubeFiles.push(fullPath); // Ya es absoluta
        }
      }
    }

    recurse(baseDir);

    // Ordenar por número si los archivos comienzan con un número
    cubeFiles.sort((a, b) => {
      const aNum = parseInt(path.basename(a));
      const bNum = parseInt(path.basename(b));
      return (isNaN(aNum) ? 0 : aNum) - (isNaN(bNum) ? 0 : bNum);
    });

    return cubeFiles;
  }
}

module.exports = FileUtils;
