// @dbcube/cli/src/utils/ui.js
const chalk = require('chalk');

/**
 * UI compartida del CLI de DBCube. Un solo lugar para el estilo visual:
 * colores de marca, símbolos de estado y el manejo de errores consistente.
 * Así todos los comandos se ven como parte del mismo producto.
 */

// Verde de marca DBCube (con fallback si la terminal no soporta truecolor).
const brandGreen = (t) => (chalk.level >= 2 ? chalk.hex('#22C55E') : chalk.green)(t);

const ui = {
    chalk,
    brand: (t) => brandGreen(chalk.bold(t)),

    /** Banner de marca para el header del CLI/help. */
    banner(subtitle = 'Database toolkit powered by Rust') {
        const line = chalk.gray('─'.repeat(54));
        console.log('');
        console.log(`  ${brandGreen('◆')} ${ui.brand('DBCube')} ${chalk.gray('CLI')}`);
        console.log(`  ${chalk.gray(subtitle)}`);
        console.log(`  ${line}`);
    },

    /** Cabecera de comando: ícono + título + línea separadora. */
    header(title, icon = '🧊') {
        console.log(`\n${icon}  ${ui.brand(title.toUpperCase())}`);
        console.log(chalk.gray('─'.repeat(54)));
    },

    /** Línea de detalle bajo una cabecera (clave: valor). */
    item(label, value) {
        if (value === undefined) {
            console.log(`${chalk.gray('│')}  ${label}`);
        } else {
            console.log(`${chalk.gray('│')}  ${chalk.gray(label + ':')} ${chalk.white(value)}`);
        }
    },

    success(message) { console.log(`${chalk.green('✔')} ${message}`); },
    info(message) { console.log(`${chalk.cyan('ℹ')} ${message}`); },
    warn(message) { console.log(`${chalk.yellow('⚠')} ${message}`); },
    dim(message) { console.log(chalk.gray(message)); },
    step(message) { console.log(`${chalk.cyan('›')} ${message}`); },

    /** Error con título en rojo y pistas opcionales (→ ...). */
    error(message, hints = []) {
        console.error(`\n${chalk.red('✖')} ${chalk.red(message)}`);
        for (const h of hints) console.error(`  ${chalk.gray('→')} ${chalk.gray(h)}`);
        console.error('');
    },

    /** "Cancelado." silencioso (Ctrl+C / prompt cerrado). */
    cancelled() {
        console.log(`\n${chalk.gray('Cancelled.')}`);
    },

    /**
     * Manejador central de errores de comando. Llamar desde cada catch.
     * Traduce los fallos comunes a mensajes claros y accionables, sin stacks
     * (salvo DBCUBE_DEBUG=1), y cierra el proceso con el código adecuado.
     */
    handleError(error) {
        // Ctrl+C / cierre del prompt interactivo.
        if (error && (error.name === 'ExitPromptError' || /force closed/i.test(error.message || ''))) {
            ui.cancelled();
            process.exit(0);
        }

        const msg = (error && error.message) || String(error);

        // Base de datos/config no inicializada (errores internos típicos).
        if (msg.includes("reading 'init'") || msg.includes("reading 'getDatabase'") || msg.includes("reading 'type'")) {
            ui.error('Database configuration not found.', [
                'Run: npx dbcube run database:create',
                'Or check that the database is configured in dbcube.config.js',
            ]);
            process.exit(1);
        }

        ui.error(msg);
        if (process.env.DBCUBE_DEBUG === '1' || process.env.DBCUBE_DEBUG === 'true') {
            console.error(error);
        }
        process.exit(1);
    },
};

module.exports = ui;
